NEP-488: Precompile for BLS12-381 curve operations (#488)
near/NEPs: neps/nep-0488.md

to their corresponding subgroups: $G_1 \subset E(F_p)$ or $G_2 \subset E'(F_{p^2})$. 
`bls12381_map_fp_to_g1`/`bls12381_map_fp2_to_g2` combine the functionalities 
of `map_to_curve` and `clear_cofactor` from RFC9380[^63].

```text
fn bls12381_map_fp_to_g1(u):
    let Q = map_to_curve(u);
    return clear_cofactor(Q);
```

We choose not to implement the `hash_to_field` function as a host function due to potential changes in hashing methods. 
Additionally, executing this function within the contract consumes approximately 2 TGas, which is acceptable for our goals.

Specific implementation parameters for `bls12381_map_fp_to_g1` and `bls12381_map_fp2_to_g2` can be found in RFC9380 
under sections 8.8.1[^64] and 8.8.2[^65], respectively.

### Curve points encoding

#### General comments

The encoding rules for curve points and field elements align with the standards established in zkcrypto[^53] and 
the implementation in the milagro lib[^29].

For elements from $F_p$ the first three bits will always be $0$, because the first byte of $p$ equals $1$. As a result, 
we can use these bits to encode extra information: the encoding format, the point at infinity, and the points' sign. 

> Read more in sections: Uncompressed/compressed points on curve $E(F_p)$/$E'(F_{p^2})$.

Reviewer: GitHub doesn't render the latter LaTeX expression `$E'(F_{p^2})$` .

Author: fixed: https://github.com/near/NEPs/pull/488/commits/52f95599df93c2608e1a7304a5b4bc54de5fcaa5

[^11]: BLS12-381 for the Rest of Us: [https://hackmd.io/@benjaminion/bls12-381](https://hackmd.io/@benjaminion/bls12-381)
[^12]: BN254 for the Rest of Us: [https://hackmd.io/@jpw/bn254](https://hackmd.io/@jpw/bn254)
[^13]: Some analytics of different curve security: [https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-02.html#name-for-100-bits-of-security](https://www.ietf.org/archive/id/draft-irtf-cfrg-pairing-friendly-curves-02.html#name-for-100-bits-of-security)
[^14]: ZCash Transfer from bn254 to bls12-381: [https://electriccoin.co/blog/new-snark-curve/](https://electriccoin.co/blog/new-snark-curve/)
[^15]: EIP-2537 Precompiles for Ethereum for BLS12-381: [https://eips.ethereum.org/EIPS/eip-2537](https://eips.ethereum.org/EIPS/eip-2537)
[^17]: Article about Rainbow Bridge [https://near.org/blog/eth-near-rainbow-bridge](https://near.org/blog/eth-near-rainbow-bridge)
[^19]: Intro into zkSNARKs: [https://media.consensys.net/introduction-to-zksnarks-with-examples-3283b554fc3b](https://media.consensys.net/introduction-to-zksnarks-with-examples-3283b554fc3b)
[^20]: Zeropool project: [https://zeropool.network/](https://zeropool.network/)

> [^24]: Precompiles on Aurora: [https://doc.aurora.dev/evm/precompiles/](https://doc.aurora.dev/evm/precompiles/)

Reviewer: The link seems to have changed to https://doc.aurora.dev/dev-reference/precompiles/

Author: fixed: https://github.com/near/NEPs/pull/488/commits/59c7f7fb286e5dc8ba35ed41a8bf5a0f77002105

---
NEP: 488
Title: Host Functions for BLS12-381 Curve Operations
Authors: Olga Kuniavskaia <olga.kunyavskaya@aurora.dev>
Status: Draft
DiscussionsTo: https://github.com/nearprotocol/neps/pull/488
Type: Runtime Spec
Version: 0.0.1
Created: 2023-07-17
LastUpdated: 2023-11-21
---

## Summary

This NEP introduces host functions to perform operations on the BLS12-381 elliptic curve. It is a minimal set of functions needed to efficiently verify BLS signatures and zkSNARKs.

## Motivation

The primary aim of this NEP is to enable fast and efficient verification of BLS signatures and zkSNARKs based on 

> BLS12-381[^1],[^11],[^52] elliptic curve on NEAR.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

To efficiently verify zkSNARKs[^19], host functions for operations on the BN254 
elliptic curve(also known as Alt-BN128)[^9], [^12] have already been implemented on NEAR[^10]. 
For instance, the Zeropool[^20] project utilizes these host functions for verifying zkSNARKs on NEAR. 
However, recent research shows that the BN254 security level is lower than 100-bit[^13] and it is not recommended for use. 
BLS12-381, on the other hand, offers over 120 bits of security[^8] and is widely used[^2],[^3],[^4],[^5],[^6],[^7]  as a robust alternative. 
Supporting operations for BLS12-381 elliptic curve will significantly enhance the security of projects similar to Zeropool.

Another crucial objective is the verification of BLS signatures. 

> Initially, host functions for BN254 in NEAR were designed for zkSNARK verification and 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

are insufficient for BLS signature verifications. 

> However, even if they were sufficient, it wouldn't work for us. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

Projects such as ZCash[^2], Ethereum[^3], Tezos[^5], and Filecoin[^6] incorporate BLS12-381 specifically within their protocols. 
If we aim for compatibility with these projects, we must also utilize this elliptic curve. 
For instance, to create a trustless bridge[^17] between Ethereum and NEAR, 
we must efficiently verify BLS signatures based on BLS12-381, as these are the signatures employed within Ethereum's protocol.

In this NEP, we propose to add the following host functions:

- ***bls12381_g1_sum —*** sum the signed points from $G_1$ on an elliptic curve. This function is useful for aggregating public keys in BLS Signature. It can be employed for simple addition in $G_1$. It is kept separate from the multiexp function due to gas cost considerations.

> - ***bls12381_g2_sum —*** sum the signed points from $G_2$ on an elliptic curve. This function is useful for aggregating signatures in BLS Signature. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_g1_multiexp —*** for points $g_i \in G_1$ and scalars $s_i$ calculate $\sum g_i s_i$. This operation can be utilized to multiply a group element by a scalar. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_g2_multiexp —*** for points $g_i \in G_2$ and scalars $s_i$ calculate $\sum g_i s_i$. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_map_fp_to_g1 —*** maps a base field elements into the $G_1$ points. It does not perform the mapping of the byte string into field elements.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_map_fp2_to_g2 —*** maps an extension field elements into the $G_2$ points. It does not perform the mapping of the byte string into extension field elements. We need this function to efficiently map a message into a group element. We are not implementing the hash_to_field[^60] function because it can be executed within a contract and various hashing algorithms can be applied.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_decompress_g1 —*** decompresses the points from $G_1$ provided in the compressed form. Certain protocols offer points on the curve in compressed form (e.g., the light client updates in Ethereum 2.0), and decompression is a time-consuming operation. All the other functions in this NEP only accept decompressed points for simplicity and optimized gas consumption.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_decompress_g2 —*** decompresses the points from $G_2$ provided in the compressed form.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

> - ***bls12381_pairing_check —*** verifying that $\prod e(p_i, q_i) = 1$, where $e$ is a pairing operation and $p_i \in G_1 \land q_i \in G_2$. This function is used to verify BLS signatures or zkSNARKs. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

Functions required for verifying BLS signatures[^59]:

- bls12381_g1_sum
- bls12381_g2_sum
- bls12381_map_fp2_to_g2
- bls12381_decompress_g1
- bls12381_decompress_g2
- bls12381_pairing_check

Functions required for verifying zkSNARKs:

- bls12381_g1_sum
- bls12381_g1_multiexp
- bls12381_pairing_check

> Both zkSNARKs and BLS signatures can be implemented alternatively by swapping G1 and G2. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

Therefore, all functions have been implemented for both G1 and G2.

An analogous proposal, EIP-2537[^15], exists in Ethereum. 
The functions here have been designed with compatibility 

> with Ethereum's proposal in mind. This design approach aims 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8e0feb93d371e4a951c49f1db020ebcf654ca770

to ensure future ease in supporting corresponding precompiles for Aurora[^24].

## Specification

### BLS12-381 Curve Specification

#### Elliptic Curve

**The field $F_p$** for some *prime* $p$ is a set of integer 
elements $\textbraceleft 0, 1, \ldots, p - 1 \textbraceright$ with two 
operations: multiplication $\cdot$ and addition $+$.
These operations involve standard integer multiplication and addition, 

> followed by taking the remainder modulo $p$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

> **The elliptic curve $E(F_p)$**  is a set of all pairs $(x, y) \in F_p$:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

$$
y^2 \equiv x^3 + Ax + B \mod p
$$

> together with an imaginary point at infinity 0, where: $A, B \in F_p$, p is prime > 3, and $4A^3 + 27B^2 \not \equiv 0 \mod p$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

In the case of BLS12-381 equation is $y^2 \equiv x^3 + 4 \mod p$[^15],[^51],[^14],[^11]

**Parameters for our case:**

- $A = 0$
- $B = 4$

> - $p = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

> Let’s $P \in E(F_q)$ have coordinates (x, y), define **$-P$**  as a point on a curve with coordinates (x, -y).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

> **The addition operation for Elliptic Curve** is a function $+\colon E(F_p) \times E(F_p) \rightarrow E(F_p)$ defined with following rules: let’s P and Q $\in E(F_p)$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

- if  $P \ne Q$ and $P \ne -Q$
  - draw a line passing through P and Q. This line intersects the curve at a third point R

>   - reflect the point R about the x-axis by changing the sign of the y-coordinate. The resulting point is P+Q.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

- if $P=Q$
  - draw a tangent line throw P for an elliptic curve. The line will intersect the curve at the second point R.
  - reflect the point R about the x-axis the same way to get point 2P
- $P = -Q$
  - $P + Q = P + (-P) = 0$ — the point on infinity
- Q = 0
  - $P + Q = P + 0 = P$

With the addition operation, Elliptic Curve forms a **group**.

#### Subgroups

**Subgroup** H is a subset of the group G with the following properties:

- $\forall h_1, h_2 \in H\colon h_1 + h_2 \in H$
- $0 \in H$
- $\forall h \in H \colon -h \in H$

Notation: $H \subseteq G$

Group/subgroup **order** is the number of elements in group/subgroup.

Notation: |G|  or #G, where G represents the group.

> For some technical reason (for `pairing` operation which we will define later),

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

> we will operate not with the entire $E(F_p)$,

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

> but only with the two subgroups $G_1$ and $G_2$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

having the same **order** $r$. 
$G_1$ is a subset of $E(F_p)$, 
while $G_2$ is a subgroup of another group that we will define later.
The value of $r$ should be a prime number, and $G1 \ne G2$

> For our BLS12-381 Elliptic Curve, **the order r** of $G1$  and $G2$[^15],[^51] is given by:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

- $r = 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001$

#### Field extension

> **The field extension $F_{p^k}$** is a set comprising all polynomials with a degree < k and coefficients from $F_p$, along with defined operations of multiplication ($\cdot$) and addition ($+$).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

$$
a_{k - 1}x^{k - 1} + \ldots + a_1x + a_0 = A(x) \in F_{p^k} \vert a_i \in F_p
$$

The addition operation ($+$) is defined as regular polynomial addition:

$$
A(x) + B(x) = C(x)
$$

$$
\sum a_i x^i + \sum b_i x^i = \sum c_i x^i
$$

$$
c_i = (a_i + b_i) \mod p
$$

The multiplication $\cdot$ is defined as regular polynomial multiplication modulo $M(x)$, 
where $M(x)$ is an irreducible polynomial of degree $k$ with coefficients from $F_p$.

$$
C(x) = A(x) \cdot B(x)\mod M(x)
$$

Notation: $F_{p^k} = F_{p}[x] / M(x)$

In BLS12-381, we will require $F_{p^{12}}$. 
We'll construct this field not directly as an extension from $F_p$, 
but rather through a stepwise process. First, we'll build $F_{p^2}$ 
as a quadratic extension of the field $F_p$. 
Second, we'll establish $F_{p^6}$ as a cubic extension of $F_{p^2}$. 

> Finally, we'll create $F_{p^{12}}$, a quadratic extension of the 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

field $F_{p^6}$.

To define these fields, we'll need to set up three irreducible polynomials[^51]:

- $F_{p^2} = F_p[u] / (u^2 + 1)$
- $F_{p^6} = F_{p^2}[v] / (v^3 - u - 1)$
- $F_{p^{12}} = F_{p^6}[w]/(w^2 - v)$

> The second subgroup we'll utilize has an order of r and 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

resides within the same elliptic curve but with elements from $F_{p^{12}}$.   
Specifically, $G_2 \subset E(F_{p^{12}})$, where $E: y^2 = x^3 + 4$

#### Twist

Storing elements from $E(F_{p^{12}})$ consumes a significant amount of memory. 
The twist operation transforms the original curve $E(F_{p^{12}})$ into another curve within a different space, 
denoted as $E'(F_{p^2})$. It is crucial that this new curve also includes a $G'_2$ subgroup with order 'r' 
so that we can easily transform it back to the original $G_2$.

We want to have $\psi \colon E'(F_{p^2}) \rightarrow E(F_{p^{12}})$, such as

- $\forall a, b \in E'(F_{p^2}) \colon \psi(a + b) = \psi(a) + \psi(b)$
- $\forall a, b \in E'(F_{p^2}) \colon \psi(a) = \psi(b) \Rightarrow a = b$

This is referred to as an injective group homomorphism.

For BLS12-381, E’ is defined as[^51]:

$$
E'\colon y^2 = x^3 + 4(u + 1)
$$

In most cases, we will be working with points from $G_2' \subset E'(F_{p^2})$ and will simply use the notation $G_2$ for this subgroup.

#### Generators

> If there exists an element g in the group G such that $\textbraceleft g, 2g, 3g, \ldots, |G|g \textbraceright = G$, the group G is called a ***cyclic group*** and g is termed a ***generator***

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

$G_1$ and $G_2$ are cyclic subgroups with the following generators[^15],[^51]:

$G_1$:

- $x = 0x17f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb$
- $y = 0x08b3f481e3aaa0f1a09e30ed741d8ae4fcf5e095d5d00af600db18cb2c04b3edd03cc744a2888ae40caa232946c5e7e1$

For $(x', y') \in G_2 \subset E'(F_{p^2}):$
$$x' = x_0 + x_1u$$

$$y' = y_0 + y_1u$$

$G_2$:

- $x_0 = 0x024aa2b2f08f0a91260805272dc51051c6e47ad4fa403b02b4510b647ae3d1770bac0326a805bbefd48056c8c121bdb8$
- $x_1 = 0x13e02b6052719f607dacd3a088274f65596bd0d09920b61ab5da61bbdc7f5049334cf11213945d57e5ac7d055d042b7e$
- $y_0 = 0x0ce5d527727d6e118cc9cdc6da2e351aadfd9baa8cbdd3a76d429a695160d12c923ac9cc3baca289e193548608b82801$
- $y_1 = 0x0606c4a02ea734cc32acd2b02bc28b99cb3e287e85a763af267492ab572e99ab3f370d275cec1da1aaa9075ff05f79be$

> **Cofactor** is the ratio of the size of the entire group G to the size of the subgroup H:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

$$
|G|/|H|
$$

Cofactor $G_1\colon h = |E(F_p)|/r$[^51]

$$h = 0x396c8c005555e1568c00aaab0000aaab$$

Cofactor $G_2\colon h' = |E'(F_{p^2})|/r$[^51]

$$h' = 0x5d543a95414e7f1091d50792876a202cd91de4547085abaa68a205b2e5a7ddfa628f1cb4d9e82ef21537e293a6691ae1616ec6e786f0c70cf1c38e31c7238e5$$

#### Pairing

> Pairing is an operation necessary for digital signatures and zkSNARKs verification. It performs the operation $e\colon G_1 \times G_2 \rightarrow G_T$, where $G_T \subset F_{p^{12}}$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

The main properties of the pairing operation are:

- $e(P, Q + R) = e(P, Q) \cdot e(P, R)$
- $e(P  + S, R) = e(P, R)\cdot e(S, R)$

To compute this function, we utilize an algorithm called Miller Loop. 

> For effective implementation of this algorithm, 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

we require a key parameter for the BLS curve, denoted as $x$:

$$ x = -0xd201000000010000$$

This parameter can be found in the following sources:

> - [^15] section specification, pairing parameters, miller loop scalar

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

- [^51] section 4.2.1 Parameter t
- [^14] section BLS12-381, parameter u
- [^11] section Curve equation and parameters, parameter x

#### Summary

The parameters for the BLS12-381 curve are as follows:

Base field modulus: $p = 0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab$

$$
E\colon y^2 \equiv x^3 + 4
$$

$$
E'\colon y^2 \equiv x^3 + 4(u + 1)
$$

Main subgroup order: $r = 0x73eda753299d7d483339d80809a1d80553bda402fffe5bfeffffffff00000001$

$$
F_{p^2} = F_p[u] / (u^2 + 1)
$$

$$
F_{p^6} = F_{p^2}[v] / (v^3 - u - 1)
$$

$$
F_{p^{12}} = F_{p^6}[w] / (w^2 - v)
$$

> Generator for G1:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

- $x = 0x17f1d3a73197d7942695638c4fa9ac0fc3688c4f9774b905a14e3a3f171bac586c55e83ff97a1aeffb3af00adb22c6bb$
- $y = 0x08b3f481e3aaa0f1a09e30ed741d8ae4fcf5e095d5d00af600db18cb2c04b3edd03cc744a2888ae40caa232946c5e7e1$

Generator for G2:

- $x_0 = 0x024aa2b2f08f0a91260805272dc51051c6e47ad4fa403b02b4510b647ae3d1770bac0326a805bbefd48056c8c121bdb8$
- $x_1 = 0x13e02b6052719f607dacd3a088274f65596bd0d09920b61ab5da61bbdc7f5049334cf11213945d57e5ac7d055d042b7e$
- $y_0 = 0x0ce5d527727d6e118cc9cdc6da2e351aadfd9baa8cbdd3a76d429a695160d12c923ac9cc3baca289e193548608b82801$
- $y_1 = 0x0606c4a02ea734cc32acd2b02bc28b99cb3e287e85a763af267492ab572e99ab3f370d275cec1da1aaa9075ff05f79be$

> Cofactor for G1:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

$$h = 0x396c8c005555e1568c00aaab0000aaab$$

> Cofactor for G2:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

$$h' = 0x5d543a95414e7f1091d50792876a202cd91de4547085abaa68a205b2e5a7ddfa628f1cb4d9e82ef21537e293a6691ae1616ec6e786f0c70cf1c38e31c7238e5$$

Key  BLS12-381 parameter used in Miller Loop:
$$x = -0xd201000000010000$$

All parameters were sourced from [^15], [^51], and [^14], and they remain consistent across these sources.

### Map to curve specification

This section delineates the functionality of the `bls12381_map_fp_to_g1` and `bls12381_map_fp2_to_g2` functions, 
operating in accordance with the RFC9380 specification "Hashing to Elliptic Curves"[^62].

> These functions map the field elements in $F_p$ or $F_{p^2}$ 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

to their corresponding subgroups: $G_1 \subset E(F_p)$ or $G_2 \subset E'(F_{p^2})$. 
`bls12381_map_fp_to_g1`/`bls12381_map_fp2_to_g2` combine the functionalities 

> of map_to_curve and clear_cofactor from RFC9380[^63].

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

```text
fn bls12381_map_fp_to_g1(u):
    let Q = map_to_curve(u);
    return clear_cofactor(Q);
```

> We choose not to implement the hash_to_field function as a host function due to potential changes in hashing methods. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/616fd038625914623677712abb0f047412f2b0d0

Additionally, executing this function within the contract consumes approximately 2 TGas, which is acceptable for our goals.

Specific implementation parameters for bls12381_map_fp_to_g1 and bls12381_map_fp2_to_g2 can be found in RFC9380 
under sections 8.8.1[^64] and 8.8.2[^65], respectively.

### Curve points encoding

#### Sign

The sign of a point on the elliptic curve is represented as a u8 type in Rust, with two possible values: 0 for a positive sign and 1 for a negative sign. Any other u8 value is considered invalid and should be treated as incorrect.

#### Scalar

> The scalar value is encoded as a little-endian [u8; 32]. All possible byte combinations are allowed.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

> #### Fields elements Fp

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

Values from $F_p$ are encoded as big-endian [u8; 48]. Only values less than p are permitted. If the value is equal to or greater than p, an error should be returned.

> #### Extension fields elements Fp2

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

> An element $q \in F_{p^{2}}$ can be expressed as $q = c_0 + c_1 v$, where $c_0, c_1 \in F_p$. The element from $F_{p^2}$ is encoded in [u8; 96] as a byte concatenation of $c_1$ and $c_0$. The encoding for $c_1$ and $c_0$ follows the rule described in the previous section.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

The representation of $q \in F_{p^2}$, with $q = c_0 + c_1 v$, is encoded as [u8; 96]:

- $c_1 \in F_p$ is represented as `[u8; 48]`
- $c_0 \in F_p$ is represented as `[u8; 48]`

#### Uncompressed points on curve E(Fp)

Points on the curve are represented by affine coordinates: $(x: F_p, y: F_p)$. 

> The elements from $E(F_p)$ are encoded in `[u8; 96]` as a byte concatenation of x and y point coordinates, where $x, y \in F_p$. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

The encoding follows the rules outlined in the section “Fields elements $F_p$”.

$E(F_p)$ is encoded as [u8; 96]:

- $x \in F_p$ is represented as [u8; 48]
- $y \in F_p$ is represented as [u8; 48]

*The second-highest bit* within the encoding serves to signify a point at infinity.
When this bit is set to 1, it designates an infinity point.
In this case, all other bits should be set to 0.

> Encoding point on infinity:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

The points on the curve are represented by affine coordinates: $(x: F_p, y: F_p)$. 
Elements from $E(F_p)$ in compressed form are encoded as `[u8; 48]`, 
with big-endian encoded $x \in F_p$. 
The $y$ coordinate is determined by the formula: $y = \pm \sqrt{x^3 + 4}$.

- The highest bit indicates if the point is encoded in compressed form and should be set to 1.
- The second-highest bit marks a point as infinity (if set to 1).
  - For infinity points, all bits except the first two should be set to 0; other encodings should be considered as incorrect.
- To represent the sign of $y$, the third-highest bit in the x encoding is utilized.
  - If the bit is 0, $y$ is positive; if 1, $y$ is negative. We'll consider the number positive by taking the smallest value between $y$ and $-y$.

The encoding for $x \in F_p$ as `[u8; 48]` bytes follows the rules described in the section "Extension fields elements $F_{p}$".

Encoding a point on $E(F_p)$ with a negative $y$ coordinate:

```rust
let x: [u8; 48] = encodeFp(x)
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x20;
```

Encoding the point of infinity:

```rust
let x: [u8; 48] = [0; 48];
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x40;
```

> #### Uncompressed points on twisted curve E'(Fp2)

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

The points on the curve are represented by affine coordinates: $(x: F_{p^2}, y: F_{p^2})$. 
Elements from $E'(F_{p^2})$ are encoded in [u8; 192] as a concatenation of bytes representing x and y coordinates, where $x, y \in F_{p^2}$. 
The encoding for $x$ and $y$ follows the rules detailed in the "Extension Fields Elements $F_{p^2}$" section.

*The second-highest bit* within the encoding serves to signify a point at infinity. 
When this bit is set to 1, it designates an infinity point. 
In this case, all other bits should be set to 0.

Encoding the point of infinity:

```bash
let x: [u8; 192] = [0; 192];
x[0] = x[0] | 0x40;
```

#### Compressed points on twisted curve E'(Fp2)

The points on the curve are represented by affine coordinates: $(x: F_{p^2}, y: F_{p^2})$. 
Elements from $E'(F_{p^2})$ in compressed form are encoded as [u8; 96], 
with big-endian encoded $x \in F_{p^2}$. 
The $y$ coordinate is determined using the formula: $y = \pm \sqrt{x^3 + 4(u + 1)}$.

- The highest bit indicates if the point is encoded in compressed form and should be set to 1.

> - The second-highest bit marks a point as infinity (if set to 1).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

>   - For infinity points, all bits except the first two should be set to 0; other encodings should be considered as incorrect.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

- To represent the sign of $y$, the third-highest bit in the x encoding is utilized.
  - If the bit is 0, $y$ is positive; if 1, $y$ is negative. We'll consider the number positive by taking the smallest value between $y$ and $-y$: first compare $c_1$, then $c_0$.

The encoding of $x \in F_{p^2}$ as [u8; 96] bytes follows the rules from the section “Extension Fields Elements $F_{p^2}$”.

Encoding a point on $E'(F_{p^2})$ with a negative $y$ coordinate:

```rust
let x: [u8; 96] = encodeFp2(x);
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x20;
```

> Encoding the point of infinity:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

```rust
let x: [u8; 96] = [0; 96];
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x40;
```

#### ERROR_CODE

Validating the input for the host functions within the contract can consume significant gas.
For instance, verifying if a point belongs to the subgroup is gas-consuming. 
If an error is returned by the near host function, the entire execution is reverted. 
To mitigate this, when the input verification is complex, the host function 
will successfully complete its work but return an ERROR_CODE. 
This enables users to handle error cases independently. It's important to note that host functions 

> might terminate with an error if it's straightforward to avoid (e.g., incorrect input size).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

The ERROR_CODE is encoded as a little-endian u64 and can hold the following values:

- 0: No error, execution was successful.
- 1: Execution finished with error due to:
  - Incorrect encoding (e.g., incorrectly set compression/decompression bit, coordinate >= p, etc.). 
  - A point not on the curve (where applicable). 
  - A point not in the expected subgroup (where applicable).

#### General comments

The encoding rules for curve points and field elements align with the standards established in zkcrypto[^53] and the implementation in the milagro lib[^29].

### Host functions

#### bls12381_g1_sum

***Description:***

The function calculates the sum of signed elements on the BLS12-381 curve. It accepts an arbitrary number of pairs $(s_i, p_i)$, where $p_i \in E(F_p)$ represents a point on the elliptic curve, and $s_i \in {0, 1}$ signifies the point's sign. The output is a single point from $E(F_p)$ equivalent to $\sum (-1)^{s_i}p_i$.

The operations, including the $E(F_p)$ curve, points on the curve, multiplication by -1, and the addition operation, are detailed in the BLS12-381 Curve Specification section.

> Note: The function accepts points from the entire curve, not restricted to $G_1$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

***Input:***

The sequence of pairs $(s_i, p_i)$, where $p_i \in E(F_p)$ represents a point and $s_i \in {0, 1}$ denotes the sign. Each point is encoded in decompressed form as $(x\colon F_p, y\colon F_p)$, and the sign is encoded in one byte, taking only two allowed values: 0 or 1. Expect 97*k bytes as input, which are interpreted as byte concatenation of k slices, with each slice representing the point sign and the uncompressed point from $E(F_p)$. Further details are available in the Curve Points Encoding section.

***Output:*** 

The ERROR_CODE is returned. 

- ERROR_CODE = 0: the input is correct
  - <ins>Output:</ins> 96 bytes represent one point $\in E(F_p)$ in its decompressed form. In case of an empty input, it outputs a point on infinity (refer to the Curve Points Encoding section for more details).
- ERROR_CODE = 1:
  - Points or signs are incorrectly encoded (refer to Curve points encoded section).
  - Point is not on the curve.

***Test cases:***

<ins>Tests for the sum of two points</ins>

> This section aims to verify the correctness of summing two valid elements on the curve:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

- Utilize points on the curve with known addition results for comparison, such as tests from EIP-2537[^47],[^48].
- Generate random points on the curve and verify the commutative property: P + Q = Q + P.

> - Validate that the sum of random points from G1 remains in G1.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

- Generate random points on the curve and use another library to cross-check the results.

Edge cases:

> - Points not from G1.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

- 0 + 0 = 0.
- P + 0 = 0 + P = P.
- P + (-P) = (-P) + P = 0.
- P + P (tangent to the curve).
- The sum of two points P and (-(P + P)) (tangent to the curve at point P).

<ins>Tests for inversion</ins>

This section aims to validate the correctness of point inversion:

- Generate random points on the curve and verify P - P = -P + P = 0.
- Generate random points on the curve and verify -(-P) = P.
- Generate random points from G1 and ensure that -P also belong to G1.
- Utilize an external implementation, generate random points on the curve, and compare results.

Edge cases:

> - Point not from G1

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

- -0

<ins>Tests for incorrect data</ins>

This section aims to validate the handling of incorrect input data:

- Incorrect input length.
- Incorrect sign value (not 0 or 1).
- Erroneous coding of field elements, resulting in a correct point on the curve if only the suffix is considered.
- Erroneous coding of field elements resulting in a correct element on the curve modulo p.
- Erroneous coding of field elements with an incorrect extra bit in the decompressed encoding.
- Point not on the curve.

> - Incorrect encoding of the point on infinity.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

- Input is beyond memory bounds.

<ins>Tests for the sum of an arbitrary amount of points</ins>

This section focuses on validating the summation functionality with an arbitrary number of points:

> - Generate random points on the curve and verify that the sum of the random permutation matches.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

- Generate random points on the curve and utilize another library to validate results.
- Create points and cross-check the outcome with the multiexp function.
- Generate random points from $G_1$ and confirm that the sum is also from $G_1$.

Edge cases:

- Empty input
- Sum with the maximum number of elements
- A single point

***Annotation:***

```rust
pub fn bls12381_g1_sum(&mut self, 
                       value_len: u64, 
                       value_ptr: u64, 
                       register_id: u64) -> Result<u64>;
```

#### bls12381_g2_sum

***Description:***

The function computes the sum of the signed elements on the BLS12-381 curve. It accepts an arbitrary number of pairs $(s_i, p_i)$, where $p_i \in E'(F_{p^2})$ represents a point on the elliptic curve and $s_i \in {0, 1}$ is the point's sign. The output is a single point from $E'(F_{p^2})$ equal to $\sum (-1)^{s_i}p_i$.

The $E'(F_{p^2})$ curve, the points on the curve, the multiplication by -1, and the addition operation are all defined in the BLS12-381 Curve Specification section.

> Note: The function accepts any points on the curve, not limited to $G_2$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

***Input:*** 

The sequence of pairs $(s_i, p_i)$, where $p_i \in E'(F_{p^2})$ is point and $s_i \in \textbraceleft 0, 1 \textbraceright$  represents a sign.  
Each point is encoded in decompressed form as $(x: F_{p^2}, y: F_{p^2})$, and the sign is encoded in one byte. The expected input size is 193*k bytes, interpreted as a byte concatenation of k slices, 
each slice representing the point sign alongside the uncompressed point from $E'(F_{p^2})$. 
More details are available in the Curve Points Encoding section.

***Output:***

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct

>   - <ins>Output:</ins> 192 bytes represent one point $\in E'(F_{p^2})$ in its decompressed form. In case of an empty input, it outputs a point on infinity (refer to the Curve Points Encoding section for more details).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

The scalar multiplication operation signifies the addition of that point a scalar number of times:

$$
s \cdot p = \underbrace{p + p + \ldots + p}_{s}
$$

The $E(F_p)$ curve, points on the curve, and the addition operation are defined in the BLS12-381 Curve Specification section.

Please note:

- The function accepts any points on the curve, not solely from $G_1$.
- The scalar is an arbitrary unsigned integer and can exceed the group order.
- To enhance gas efficiency, the Pippenger’s algorithm[^25] can be utilized.

***Input:*** The sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ represents a point on the curve, and $s_i \in \mathbb{N}_0$ is a scalar. The expected input size is 128*k bytes, interpreted as byte concatenation of k slices. Each slice comprises the concatenation of an uncompressed point from $E(F_p)$— 96 bytes, along with a scalar— 32 bytes. Further details are available in the Curve Points Encoding section.

***Output:***

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct

>   - <ins>Output:</ins> 96 bytes represent one point $\in E(F_p)$ in its decompressed form. In case of an empty input, it outputs a point on infinity (refer to the Curve Points Encoding section for more details).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

The same test cases as those in the `bls12381_g1_sum` section will be applied.

***Annotation:***

```rust
pub fn bls12381_g1_multiexp(
        &mut self,
        value_len: u64,
        value_ptr: u64,
        register_id: u64,
) -> Result<u64>;
```

#### ***bls12381_g2_multiexp***

***Description:***

The function takes a list of pairs $(p_i, s_i)$ as input, where $p_i \in E'(F_{p^2})$ represents a point on the curve, and $s_i \in \mathbb{N}_0$ denotes a scalar. The function computes $\sum s_i \cdot p_i$.

This scalar multiplication operation involves adding the point $p$ to itself a specified number of times:

$$
s \cdot p = \underbrace{p + p + \ldots + p}_{s}
$$

The $E'(F_{p^2})$ curve, points on the curve, and the addition operation are defined in the BLS12-381 Curve Specification section.

Please note:

- The function accepts any points on the curve, not solely from $G_2$.
- The scalar is an arbitrary unsigned integer and can exceed the group order.
- To enhance gas efficiency, the Pippenger’s algorithm[^25] can be utilized.

***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E'(F_{p^2})$ is a point on the curve and $s_i \in \mathbb{N}_0$ is a scalar. 

> Expected `224*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the concatenation of uncompressed point from $E'(F_{p^2})$ — `192` bytes and scalar — `32` bytes. More details are in the Curve Points Encoding section.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

***Output:*** 

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct

>   - <ins>Output:</ins> 192 bytes represent one point $\in E'(F_{p^2})$ in its decompressed form. In case of an empty input, it outputs a point on infinity (refer to the Curve Points Encoding section for more details).

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

) -> Result<u64>;
```

#### bls12381_map_fp2_to_g2

***Description:***

The function takes the list of elements $a_i \in F_{p^2}$ and maps them to $G_2 \subset E'(F_{p^2})$. You can find the mapping function's specifications in the "Map to Curve Specification" section. It's important to note that this function does NOT map the byte string into $F_{p^2}$. The implementation of this function may vary and can be effectively executed within the contract.

***Input:*** the function takes as input `96*k` bytes — the elements from $F_{p^2}$ (two unsigned integers $< p$). Additional details can be found in the Curve Points Encoding section.

***Output:*** 

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct
  - <ins>Output:</ins> `192*k` bytes - represents a list of points $\in G_2 \subset E'(F_{p^2})$ in decompressed format. More details are in the Curve Points Encoding section.
- ERROR_CODE = 1: one of the values is not a valid extension field $F_{p^2}$ element

***Test cases:***

<ins>Tests for general cases </ins>

- Validate the results for known answers from EIP-2537[^47],[^48]
- Generate a random point $a$ from $F_{p^2}$:
  - Verify the result with another library.

>   - Check that result point on curve in G2.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

  - Compare results for $a$ and $-a$; they should have the same x-coordinates and opposite y-coordinates.

Edge cases:

- $a = (0, 0)$
- $a = (p - 1, p - 1)$

<ins>Tests for an arbitrary number of elements</ins>

- Empty input
- Maximum number of points.
- Generate a random number of field elements and compare the result with another library.

<ins>Tests for error cases </ins>

- Input length is not divisible by 96.
- Input is beyond memory bounds.
- $a = (0, p)$
- $a = (p, 0)$
- (random number $\ge p$, 0)
- (0, random number $\ge p$)

***Annotation:***

```rust
pub fn bls12381_map_fp2_to_g2(
        &mut self,
        value_len: u64,
        value_ptr: u64,
        register_id: u64,
) -> Result<u64>;
```

#### bls12381_pairing_check

***Description:***

The pairing function $e\colon G_1 \times G_2 \rightarrow G_T$, where $G_T \subset F_{q^{12}}$, exhibits the following properties:

- $e(P, Q + R) = e(P, Q) \cdot e(P, R)$
- $e(P + S, R) = e(P, R)\cdot e(S, R)$

Consequently:

$e([a]P, [b]Q) = e(P, Q)^{ab} = e([b]P, [a]Q)$

This function is necessary to verify BLS signatures/zkSNARKs.

> This function takes as input the sequence of pairs $(p_i, q_i)$, where $p_i \in G_1 \subset E(F_{p})$ and $q_i \in G_2 \subset E'(F_{p^2})$ and validate:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

$$
\prod e(p_i, q_i) = 1
$$

> We don’t calculate the pairing function itself: the result will be in the huge field, and in all known applications only this validation check is necessary.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

  - $e(s_1P, s_2Q) \cdot e(-s_2P, s_1Q) = 1$
  - $e(s_1P, s_2Q) \cdot e(s_2P, -s_1Q) = 1$

- $g_1 \in G_1$, $g_2 \in G_2$ are generators defined in section 'BLS12-381 Curve Specification', r is the order of $G_1$ and $G_2$,  and $p_1, p_2, q_1, q_2$ are randomly generated scalars:
  - if $p_1 \cdot q_1 + p_2 \cdot q_2 \not \equiv 0 (\mod r)$, verify $e(p_1 g_1, q_1 g_2) \cdot e(p_2 g_1, q_2 g_2) \ne 1$
  - if $p_1 \cdot q_1 + p_2 \cdot q_2 \equiv 0 (\mod r)$, verify $e(p_1 g_1, q_1 g_2) \cdot e(p_2 g_1, q_2 g_2) = 1$

<ins>Tests for an arbitrary number of pairs</ins>

- Empty input
- Test with the maximum number of pairs
- Tests using known answers from EIP-2537[^47],[^48]
- For all possible values of 'n', generate random scalars $p_1 \cdots p_n$ and $q_1 \cdots q_n$ such that $\sum p_i \cdot q_i \not \equiv 0 (\mod r)$:
  - Verify $\prod e(p_i g_1, q_i g_2) \ne 1$
- For all possible values of 'n', generate random scalars $p_1 \cdots p_{n - 1}$ and $q_1 \cdots q_{n - 1}$:
  - Verify $(\prod e(p_i g_1, q_i g_2)) \cdot e(-(\sum p_i q_i) g_1, g_2) = 1$
  - Verify $(\prod e(p_i g_1, q_i g_2)) \cdot e(g_1, -(\sum p_i q_i) g_2) = 1$

<ins>Tests for error cases</ins>

> - The first point is on the curve but not in G1.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

> - The second point is on the curve but not in G2.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

- The input length is not divisible by 288.
- The first point is not on the curve.
- The second point is not on the curve.
- Input length exceeds the memory limit.

> - Incorrect encoding of the infinity point.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f6402c45d1345f7f5d02e7b757d92e0016ead648

***Input:*** A sequence of points $p_i \in E(F_p)$, with each point encoded in compressed form. An expected input size of 48*k bytes is anticipated, interpreted as the byte concatenation of k slices. Each slice represents the compressed point from $E(F_p)$. Additional details can be found in the Curve Points Encoding section.

***Output:***

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct
  - <ins>Output:</ins> The sequence of points $p_i \in E(F_p)$, with each point encoded in decompressed form. An expected output of 96*k bytes, interpreted as the byte concatenation of k slices. Each slice represents the decompressed point from $E(F_p)$. k is the same as in the input. More details are available in the Curve Points Encoding section.
- ERROR_CODE = 1:
  - Points are incorrectly encoded (refer to the Curve points encoded section).
  - Point is not on the curve.

***Test cases:***

<ins>Tests for decompressing a single point</ins>

> - Generate random points on the curve from G1 and not from G1:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

In addition, there are implementations in other languages that are less relevant to us in this context but can serve as references.

1. C++, ETH2.0 Client, ***Chia library***[^37]
2. Haskell, ***Adjoint Lib***[^38]
3. Go, ***Go-Ethereum***[^39]
4. JavaScript, ***Noble JS***[^40]
5. Go, ***Matter Labs Go EIP-1962 implementation***[^41]
6. C++, ***Matter Labs Go EIP-1962 implementation***[^42]

One of the possible libraries to use is the blst library[^30]. 
This library exhibits good performance[^45] and has undergone several audits[^55]. 
You can find the draft implementation in nearcore, which is based on this library, through this link[^54].

## Security Implications

The implementation's security depends on the chosen library's security, supporting operations with BLS curves.

Within this NEP, a constant execution time for all operations isn't mandated. This isn't an issue when employing host functions for BLS signature/zkSNARKs verification. However, refrain from using these host functions if a constant-time algorithm is necessary.

> BLS12-381 offers more security bits compared to the already existing pairing-friendly curve BN254. Consequently, the security of projects requiring the pairing-friendly curve will be enhanced.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/32798b21bcd8e466498097b365acc5053eba15bb

## Alternatives

In nearcore, host functions for another pairing-friendly curve, BN254, have already been implemented[^10]. Some projects[^20] might consider utilizing the supported curve as an alternative. However, recent research indicates that this curve provides less than 100 bits of security and is not recommended for use[^13]. Furthermore, projects involved in cross-chain interactions, like Rainbow Bridge, are mandated to employ the same curve as the target protocol, which, in the case of Ethereum, is currently BLS12-381[^3]. Consequently, there is no viable alternative to employing a different pairing-friendly curve.

> An alternative approach involves creating a single straightforward host function in nearcore for BLS signature verification. This was the initially proposed solution[^26]. However, this solution lacks flexibility[^28] for several reasons: (1) projects may utilize different hash functions; (2) some projects might employ the G1 subgroup for public keys, while others use G2; (3) the specifications for Ethereum 2.0 remain in draft, subject to potential changes; (4) instead of a more varied and adaptable set of functions (inspired by EIP-2537's precompiles), we are left with a single large function; (5) there will be no support for zkSNARKs verification.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/c70a134b493f189b32b610206a2051ba0522a05a

Another alternative is to perform BLS12-381 operations off-chain. In this scenario, applications utilizing the BLS curve will no longer maintain trustlessness.

## Future possibilities

In the future, there might be support for working with various curves beyond just BLS12-381. In Ethereum, prior to EIP-2537[^15], there was a proposal, EIP-1962[^27], to introduce pairing-friendly elliptic curves in a versatile format, accommodating not only BLS curves but numerous others as well. However, this proposal wasn't adopted due to its extensive scope and complexity. Implementing every conceivable curve might not be practical, but it remains a potential extension worth considering.

> Another potential extension could involve supporting hash_to_field or hash_to_curve operations[^58]. Enabling their support would optimize gas usage for encoding messages into elements on the curve, which could benefit BLS signatures. However, implementing the hash_to_field operation requires supporting multiple hashing algorithms simultaneously and doesn't demand a significant amount of gas for implementation within the contract. Therefore, these functions exceed the scope of this proposal.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/32798b21bcd8e466498097b365acc5053eba15bb

Second, we'll establish $F_{p^6}$ as a cubic extension of $F_{p^2}$. 
Finally, we'll create $F_{p^{12}}$ as a quadratic extension of the 
field $F_{p^6}$.

To define these fields, we'll need to set up three irreducible polynomials[^51]:

- $F_{p^2} = F_p[u] / (u^2 + 1)$
- $F_{p^6} = F_{p^2}[v] / (v^3 - u - 1)$
- $F_{p^{12}} = F_{p^6}[w] / (w^2 - v)$

The second subgroup we'll utilize has order r and 
resides within the same elliptic curve but with elements from $F_{p^{12}}$.   
Specifically, $G_2 \subset E(F_{p^{12}})$, where $E: y^2 = x^3 + 4$

#### Twist

Storing elements from $E(F_{p^{12}})$ consumes a significant amount of memory. 
The twist operation transforms the original curve $E(F_{p^{12}})$ into another curve within a different space, 
denoted as $E'(F_{p^2})$. It is crucial that this new curve also includes a $G'_2$ subgroup with order 'r' 
so that we can easily transform it back to the original $G_2$.

We want to have $\psi \colon E'(F_{p^2}) \rightarrow E(F_{p^{12}})$, such as

- $\forall a, b \in E'(F_{p^2}) \colon \psi(a + b) = \psi(a) + \psi(b)$
- $\forall a, b \in E'(F_{p^2}) \colon \psi(a) = \psi(b) \Rightarrow a = b$

> This is referred to as an injective group homomorphism.

Author: `G'_2` and `G_2` have the same order, not `E'(Fp2)` and `E(Fp12)`. 

My intuition tells me that `E(Fp12)` is bigger then `E'(Fp2)`. At least for me it is not obviuse that `E(Fp12)` and `E'(Fp2)` have the same order. I feel like it is not true. But I don't want spent time right now for prooving.

> This is referred to as an injective group homomorphism.

Reviewer 2: SGTM! It’s not a big deal anyway, whichever way it is :)

under sections 8.8.1[^64] and 8.8.2[^65], respectively.

### Curve points encoding

#### Sign

The sign of a point on the elliptic curve is represented as a u8 type in Rust, with two possible values: 0 for a positive sign and 1 for a negative sign. Any other u8 value is considered invalid and should be treated as incorrect.

#### Scalar

A scalar value is encoded as little-endian [u8; 32]. All possible byte combinations are allowed.

#### Fields elements $F_p$

Values from $F_p$ are encoded as big-endian [u8; 48]. Only values less than p are permitted. If the value is equal to or greater than p, an error should be returned.

#### Extension fields elements $F_{p^2}$

An element $q \in F_{p^{2}}$ can be expressed as $q = c_0 + c_1 v$, where $c_0, c_1 \in F_p$. 
An element from $F_{p^2}$ is encoded in [u8; 96] as the byte concatenation of $c_1$ and $c_0$. The encoding for $c_1$ and $c_0$ follows the rule described in the previous section.

The representation of $q \in F_{p^2}$, with $q = c_0 + c_1 v$, is encoded as [u8; 96]:

- $c_1 \in F_p$ is represented as `[u8; 48]`

> - $c_0 \in F_p$ is represented as `[u8; 48]`

Author: fixed: https://github.com/near/NEPs/pull/488/commits/83a17d86b22e9e6a443b42a3b7edea70cbfcd6ab

#### Uncompressed points on curve $E(F_p)$

Points on the curve are represented by affine coordinates: $(x: F_p, y: F_p)$. 
Elements from $E(F_p)$ are encoded in `[u8; 96]` as the byte concatenation of the x and y point coordinates, where $x, y \in F_p$. 
The encoding follows the rules outlined in the section “Fields elements $F_p$”.

$E(F_p)$ is encoded as [u8; 96]:

- $x \in F_p$ is represented as [u8; 48]

> - $y \in F_p$ is represented as [u8; 48]

Author: fixed: https://github.com/near/NEPs/pull/488/commits/83a17d86b22e9e6a443b42a3b7edea70cbfcd6ab

*The second-highest bit* within the encoding serves to signify a point at infinity.

> When this bit is set to 1, it designates an infinity point.

Author: I use encoding from this doc: https://github.com/zkcrypto/pairing/blob/0.14.0/src/bls12_381/README.md

This encoding is also used in popular libraries, such as blst and milagro. Ethereum uses the same method for encoding the public keys and signatures. 

$p = \mathtt{0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab}$
p starts with 1, as a result, the first three bits will always be 0 if we will take the elements by modulo p. So, we can use these bits as we want. 

Did I answer your question?

> When this bit is set to 1, it designates an infinity point.

Reviewer 2: Yes! It’d be great to just write this down as part of the NEP, so that it’s obvious for any future reader :)

> When this bit is set to 1, it designates an infinity point.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/177d508ffac4b055eceb00ba47a28117dab2d1fe

In this case, all other bits should be set to 0.

Encoding the point at infinity:

```bash
let x: [u8; 96] = [0; 96];
x[0] = x[0] | 0x40;
```

#### Compressed points on curve $E(F_p)$

The points on the curve are represented by affine coordinates: $(x: F_p, y: F_p)$. 
Elements from $E(F_p)$ in compressed form are encoded as `[u8; 48]`, 
with big-endian encoded $x \in F_p$. 
The $y$ coordinate is determined by the formula: $y = \pm \sqrt{x^3 + 4}$.

- The highest bit indicates if the point is encoded in compressed form and should be set to 1.
- The second-highest bit marks the point at infinity (if set to 1).
  - For the point at infinity, all bits except the first two should be set to 0; other encodings should be considered as incorrect.
- To represent the sign of $y$, the third-highest bit in the x encoding is utilized.

>   - If the bit is 0, $y$ is positive; if 1, $y$ is negative. We'll consider the number positive by taking the smallest value between $y$ and $-y$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/d508b6fd3976bde6f5da142a7d097a0921161995 ?

The encoding for $x$ and $y$ follows the rules detailed in the "Extension Fields Elements $F_{p^2}$" section.

*The second-highest bit* within the encoding serves to signify a point at infinity. 
When this bit is set to 1, it designates an infinity point. 
In this case, all other bits should be set to 0.

Encoding the point at infinity:

```bash
let x: [u8; 192] = [0; 192];
x[0] = x[0] | 0x40;
```

#### Compressed points on twisted curve $E'(F_{p^2})$

The points on the curve are represented by affine coordinates: $(x: F_{p^2}, y: F_{p^2})$. 
Elements from $E'(F_{p^2})$ in compressed form are encoded as [u8; 96], 
with big-endian encoded $x \in F_{p^2}$. 
The $y$ coordinate is determined using the formula: $y = \pm \sqrt{x^3 + 4(u + 1)}$.

- The highest bit indicates if the point is encoded in compressed form and should be set to 1.
- The second-highest bit marks the point at infinity (if set to 1).
  - For the point at infinity, all bits except the first two should be set to 0; other encodings should be considered as incorrect.
- To represent the sign of $y$, the third-highest bit in the x encoding is utilized.

>   - If the bit is 0, $y$ is positive; if 1, $y$ is negative. We'll consider the number positive by taking the smallest value between $y$ and $-y$: first compare $c_1$, then $c_0$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/a634505e98b74d6e75dac1a23cf362e049185da4

x[0] = x[0] | 0x20;
```

Encoding the point at infinity:

```rust
let x: [u8; 96] = [0; 96];
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x40;
```

#### ERROR_CODE

Validating the input for the host functions within the contract can consume significant gas.
For instance, verifying if a point belongs to the subgroup is gas-consuming. 
If an error is returned by the near host function, the entire execution is reverted. 
To mitigate this, when the input verification is complex, the host function 
will successfully complete its work but return an ERROR_CODE. 
This enables users to handle error cases independently. It's important to note that host functions
might terminate with an error if it's straightforward to avoid it (e.g., incorrect input size).

The ERROR_CODE is an u64 and can hold the following values:

- 0: No error, execution was successful. For `bls12381_pairing_check` function, the pairing result equals the multiplicative identity.
- 1: Execution finished with error due to:
  - Incorrect encoding (e.g., incorrectly set compression/decompression bit, coordinate >= p, etc.). 
  - A point not on the curve (where applicable). 
  - A point not in the expected subgroup (where applicable).

> - 2: Can be returned only in `bls12381_pairing_check`. No error, execution was successful, but the pairing result doesn't equal the multiplicative identity. 

Author: I don't understand this comment. Why we should be consistent with Linux errno? The linux errno can be negative? We can not use the negative numbers because we're returning the unsigned type.

> - 2: Can be returned only in `bls12381_pairing_check`. No error, execution was successful, but the pairing result doesn't equal the multiplicative identity. 

Reviewer 2: Linux typically uses functions that return a positive number (often 1, sometimes other values) for success, and a negative number for failure, where which exact negative number defines the details of the error.

This is good because it extends seamlessly to cases where we also need to return either an error or a number between [0, i64::MAX].

So I’d suggest returning `i32` instead of `u32`, and renumbering the return values to be -1 for error, 0 for successful execution; and, for bls12381_pairing_check, 1 for "the result didn’t match".

Incidentally, `-1 as u32` is 0xFFFFFFFF, which means we could easily change the spec later on, if we were to need a function that can error with return values that go up to 0xFFFFFFFE.

> - 2: Can be returned only in `bls12381_pairing_check`. No error, execution was successful, but the pairing result doesn't equal the multiplicative identity. 

Author: In nearcore only two returned types are supported: u64 and u32. It is not possible to return the signed type. We can indeed represent the negative numbers in unsigned types, as you showed, but I am afraid that it can make the logic even more complicated. 

Your suggestion makes a lot of sense, and it is logical when we use the negative numbers for error cases and the positive ones for the results. But in this specific case, I prefer not to change anything to avoid casting between signed and unsigned types.

> - 2: Can be returned only in `bls12381_pairing_check`. No error, execution was successful, but the pairing result doesn't equal the multiplicative identity. 

Reviewer 2: > In nearcore only two returned types are supported: u64 and u32.

Near-vm (or basically any spec-compliant wasm interpreter) should work without any issue with `i32` parameters, so long as the host function is properly defined with the right types on both sides. For near-vm (or any wasm interpreter, as per the wasm spec) the parameter is just "a 32-bit value", and how to interpret it is just convention on both sides, with no preference toward one or the other.

Now, maybe I missed something, but I’d be very surprised if we were unable to have host functions that take/return `i32`

(That said, I totally understand your reluctance now! If we’re indeed unable to return `i32` then the added complexity would not make sense for now)

***Test cases:***

<ins>Tests for general cases </ins>

- Validate the results for known answers from EIP-2537[^47],[^48]
- Generate a random point $a$ from $F_{p^2}$:
  - Verify the result with another library.
  - Check that the resulting point lies in $G_2$.
  - Compare results for $a$ and $-a$; they should have the same x-coordinates and opposite y-coordinates.

Edge cases:

- $a = (0, 0)$
- $a = (p - 1, p - 1)$

<ins>Tests for an arbitrary number of elements</ins>

- Empty input
- Maximum number of points.
- Generate a random number of field elements and compare the result with another library.

<ins>Tests for error cases </ins>

- Input length is not divisible by 96.
- Input is beyond memory bounds.
- $a = (0, p)$
- $a = (p, 0)$
- (random number $\ge p$, 0)
- (0, random number $\ge p$)

***Annotation:***

```rust
pub fn bls12381_map_fp2_to_g2(
        &mut self,
        value_len: u64,
        value_ptr: u64,
        register_id: u64,
) -> Result<u64>;
```

#### bls12381_pairing_check

***Description:***

The pairing function $e\colon G_1 \times G_2 \rightarrow G_T$, where $G_T \subset F_{q^{12}}$, exhibits the following properties:

- $e(P, Q + R) = e(P, Q) \cdot e(P, R)$

> - $e(P + S, R) = e(P, R)\cdot e(S, R)$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/6f9d3a45b112ca85f7de439c0dcb14c27f055abe

- Test with the maximum number of pairs
- Tests using known answers from EIP-2537[^47],[^48]
- For all possible values of 'n', generate random scalars $p_1 \cdots p_n$ and $q_1 \cdots q_n$ such that $\sum p_i \cdot q_i \not \equiv 0 (\mod r)$:
  - Verify $\prod e(p_i g_1, q_i g_2) \ne 1$
- For all possible values of 'n', generate random scalars $p_1 \cdots p_{n - 1}$ and $q_1 \cdots q_{n - 1}$:
  - Verify $(\prod e(p_i g_1, q_i g_2)) \cdot e(-(\sum p_i q_i) g_1, g_2) = 1$
  - Verify $(\prod e(p_i g_1, q_i g_2)) \cdot e(g_1, -(\sum p_i q_i) g_2) = 1$

<ins>Tests for error cases</ins>

- The first point is on the curve but not in $G_1$.
- The second point is on the curve but not in $G_2$.
- The input length is not divisible by 288.
- The first point is not on the curve.
- The second point is not on the curve.
- Input length exceeds the memory limit.
- Incorrect encoding of the point at infinity.
- Incorrect encoding of a curve point:
  - Incorrect decompression bit.
  - Coordinates greater than or equal to 'p'.

***Annotation:***

```rust
pub fn bls12381_pairing_check(&mut self, 
                              value_len: u64, 
                              value_ptr: u64, 

>                               register_id: u64) -> Result<u64>;

Author: fixed: https://github.com/near/NEPs/pull/488/commits/f9343c85ae2d5932d95f5ae4d8ac38f4a28ca8e8

***Output:***

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct
  - <ins>Output:</ins> the sequence of point $p_i \in E'(F_{p^2})$, with each point encoded in decompressed form. The expected output is 192*k bytes, interpreted as the byte concatenation of k slices. `k` corresponds to the value specified in the input section. Each slice represents the decompressed point from $E'(F_{p^2})$. For more details, refer to the Curve Points Encoding section.
- ERROR_CODE = 1:
  - Points are incorrectly encoded (refer to Curve points encoded section). 
  - Point is not on the curve.

***Test cases:***

The same test cases as `bls12381_p1_decompress`, but with points from $G_2$, and the input length should be divisible by 96.

***Annotation:***

```rust
pub fn bls12381_p2_decompress(&mut self, 
                              value_len: u64, 
                              value_ptr: u64, 
                              register_id: u64) -> Result<u64>;
```

> #### General comments for all functions

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ebe96e2bd9f5e9a731dbe701507b4760b5726e5

4. ***zCash origin implementation*** [^33]
5. ***MCL Library*** [^34]
6. ***FileCoin implementation*** [^35]
7. ***zkCrypto*** [^36]

To compile the list, we used links from EIP-2537[^43], the pairing-curves specification[^44], and an article containing benchmarks[^45]. This list might be incomplete, but it should encompass the primary BLS12-381 implementations.

In addition, there are implementations in other languages that are less relevant to us in this context but can serve as references.

1. C++, ETH2.0 Client, ***Chia library***[^37]
2. Haskell, ***Adjoint Lib***[^38]
3. Go, ***Go-Ethereum***[^39]
4. JavaScript, ***Noble JS***[^40]
5. Go, ***Matter Labs Go EIP-1962 implementation***[^41]
6. C++, ***Matter Labs Go EIP-1962 implementation***[^42]

One of the possible libraries to use is the blst library[^30]. 
This library exhibits good performance[^45] and has undergone several audits[^55]. 
You can find a draft implementation in nearcore, which is based on this library, through this link[^54].

## Security Implications

The implementation's security depends on the chosen library's security, supporting operations with BLS curves.

> Within this NEP, a constant execution time for all operations isn't mandated. This isn't an issue when employing host functions for BLS signature/zkSNARKs verification. However, refrain from using these host functions if a constant-time algorithm is necessary.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/60edc7fceadded08640114fa62ebf9c599ab5763

[^43]: EIP-2537 with links: [https://github.com/matter-labs-forks/EIPs/blob/bls12_381/EIPS/eip-2537.md](https://github.com/matter-labs-forks/EIPs/blob/bls12_381/EIPS/eip-2537.md)
[^44]: Pairing-friendly curves specification, crypto libs: [https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-09#name-cryptographic-libraries](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-09#name-cryptographic-libraries)
[^45]: Comparing different libs for pairing-friendly curves: [https://hackmd.io/@gnark/eccbench](https://hackmd.io/@gnark/eccbench)
[^46]: Bench vectors from EIP2537: [https://eips.ethereum.org/assets/eip-2537/bench_vectors](https://eips.ethereum.org/assets/eip-2537/bench_vectors)
[^47]: Metter Labs tests for EIP2537: [https://github.com/matter-labs/eip1962/tree/master/src/test/test_vectors/eip2537](https://github.com/matter-labs/eip1962/tree/master/src/test/test_vectors/eip2537)
[^48]: Tests from Go Ethereum implementation: [https://github.com/ethereum/go-ethereum/tree/master/core/vm/testdata/precompiles](https://github.com/ethereum/go-ethereum/tree/master/core/vm/testdata/precompiles)

> [^51]: draft-irtf-cfrg-pairing-friendly-curves-11 [https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-11#name-bls-curves-for-the-128-bit-](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-pairing-friendly-curves-11#name-bls-curves-for-the-128-bit-)  *(*[https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/bls.md](https://github.com/ethereum/consensus-specs/blob/dev/specs/altair/bls.md) → [https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04) → this ref*)*

Author: fixed: https://github.com/near/NEPs/pull/488/commits/ebbf76819108b4cf76daf0fbcd99c3915a249373

---
NEP: 488
Title: Host Functions for BLS12-381 Curve Operations
Authors: Olga Kuniavskaia <olga.kunyavskaya@aurora.dev>
Status: Draft
DiscussionsTo: https://github.com/nearprotocol/neps/pull/488
Type: Runtime Spec
Version: 0.0.1
Created: 2023-07-17
LastUpdated: 2023-11-21
---

## Summary

This NEP introduces host functions to perform operations on the BLS12-381 elliptic curve. It is a minimal set of functions needed to efficiently verify BLS signatures and zkSNARKs.

## Motivation

The primary aim of this NEP is to enable fast and efficient verification of BLS signatures and zkSNARKs based on the BLS12-381[^1],[^11],[^52] elliptic curve on NEAR.

To efficiently verify zkSNARKs[^19], host functions for operations on the BN254 
elliptic curve (also known as Alt-BN128)[^9], [^12] have already been implemented on NEAR[^10]. 
For instance, the Zeropool[^20] project utilizes these host functions for verifying zkSNARKs on NEAR. 
However, recent research shows that the BN254 security level is lower than 100-bit[^13] and it is not recommended for use. 

> BLS12-381, on the other hand, offers over 120 bits of security[^8] and is widely used[^2],[^3],[^4],[^5],[^6],[^7]  as a robust alternative. 

Reviewer 3: I think that commas shouldn't be used between multiple consecutive references (for example, Wikipedia doesn't use them).

Author: Sure. But I couldn't beat the automatic style checks in git. If you do not put commas, then it will think that everything except the first link is not used, and will be unhappy with unused references. That's why I preferred to put commas so that there would be a green check mark in git.

Supporting operations for BLS12-381 elliptic curve will significantly enhance the security of projects similar to Zeropool.

Another crucial objective is the verification of BLS signatures. 
Initially, host functions for BN254 on NEAR were designed for zkSNARK verification and 
are insufficient for BLS signature verification.
However, even if these host functions were sufficient for BLS signature verification on the BN254 elliptic curve, this would not be enough for compatibility with other projects.
In particular, projects such as ZCash[^2], Ethereum[^3], Tezos[^5], and Filecoin[^6] incorporate BLS12-381 specifically within their protocols. 
If we aim for compatibility with these projects, we must also utilize this elliptic curve. 
For instance, to create a trustless bridge[^17] between Ethereum and NEAR, 
we must efficiently verify BLS signatures based on BLS12-381, as these are the signatures employed within Ethereum's protocol.

In this NEP, we propose to add the following host functions:

> - ***bls12381_g1_sum —*** computes the sum of signed points from $G_1$ on an elliptic curve. This function is useful for aggregating public keys in the BLS signature scheme. It can be employed for simple addition in $G_1$. It is kept separate from the `multiexp` function due to gas cost considerations.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/b9392075b467469e1b1cd80f21ff173020961aba

- ***bls12381_g2_multiexp —*** calculates $\sum g_i s_i$ for points $g_i \in G_2$ and scalars $s_i$. 
- ***bls12381_map_fp_to_g1 —*** maps base field elements into $G_1$ points. It does not perform the mapping of byte strings into field elements.
- ***bls12381_map_fp2_to_g2 —*** maps extension field elements into $G_2$ points.  This function does not perform the mapping of byte strings into extension field elements, which would be needed to efficiently map a message into a group element. We are not implementing the `hash_to_field`[^60] function because the latter can be executed within a contract and various hashing algorithms can be used within this function.
- ***bls12381_decompress_g1 —*** decompresses points from $G_1$ provided in a compressed form. Certain protocols offer points on the curve in a compressed form (e.g., the light client updates in Ethereum 2.0), and decompression is a time-consuming operation. All the other functions in this NEP only accept decompressed points for simplicity and optimized gas consumption.
- ***bls12381_decompress_g2 —*** decompresses points from $G_2$ provided in a compressed form.

> - ***bls12381_pairing_check —*** verifies that $\prod e(p_i, q_i) = 1$, where $e$ is a pairing operation and $p_i \in G_1 \land q_i \in G_2$. This function is used to verify BLS signatures or zkSNARKs. 

Author: fixed: https://github.com/near/NEPs/pull/488/commits/b279616b3fe9916fba6c23c3e10b0c1e149b4423

- bls12381_g2_sum
- bls12381_map_fp2_to_g2
- bls12381_decompress_g1
- bls12381_decompress_g2
- bls12381_pairing_check

Functions required for verifying zkSNARKs:

- bls12381_g1_sum
- bls12381_g1_multiexp
- bls12381_pairing_check

Both zkSNARKs and BLS signatures can be implemented alternatively by swapping $G_1$ and $G_2$. 
Therefore, all functions have been implemented for both $G_1$ and $G_2$.

An analogous proposal, EIP-2537[^15], exists in Ethereum. 
The functions here have been designed with compatibility 
with that Ethereum's proposal in mind. This design approach aims 
to ensure future ease in supporting corresponding precompiles for Aurora[^24].

## Specification

### BLS12-381 Curve Specification

#### Elliptic Curve

**The field $F_p$** for some *prime* $p$ is a set of integer 
elements $\textbraceleft 0, 1, \ldots, p - 1 \textbraceright$ with two 
operations: multiplication $\cdot$ and addition $+$.
These operations involve standard integer multiplication and addition, 
followed by computing the remainder modulo $p$.

**The elliptic curve $E(F_p)$** is the set of all pairs $(x, y)$ with coordinates in $F_p$ satisfying:

$$
y^2 \equiv x^3 + Ax + B \mod p
$$

> together with an imaginary point at infinity $\mathcal{O}$, where: $A, B \in F_p$, p is prime > 3, and $4A^3 + 27B^2 \not \equiv 0 \mod p$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

> In the case of BLS12-381 equation is $y^2 \equiv x^3 + 4 \mod p$[^15],[^51],[^14],[^11]

Reviewer 3: ```suggestion
In the case of BLS12-381 the equation is $y^2 \equiv x^3 + 4 \mod p$[^15],[^51],[^14],[^11]
```

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

**Parameters for our case:**

- $A = 0$
- $B = 4$
- $p = \mathtt{0x1a0111ea397fe69a4b1ba7b6434bacd764774b84f38512bf6730d2a0f6b0f6241eabfffeb153ffffb9feffffffffaaab}$

> Let $P \in E(F_q)$ have coordinates (x, y), define **$-P$**  as a point on a curve with coordinates (x, -y).

Reviewer 3: ```suggestion
Let $P \in E(F_q)$ have coordinates $(x, y)$, define **$-P$** as a point on a curve with coordinates $(x, -y)$.
```

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

> **The addition operation for Elliptic Curve** is a function $+\colon E(F_p) \times E(F_p) \rightarrow E(F_p)$ defined with following rules: let P and Q $\in E(F_p)$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

- if  $P \ne Q$ and $P \ne -Q$

>   - draw a line passing through P and Q. This line intersects the curve at a third point R

Reviewer 3: ```suggestion
  - draw a line passing through $P$ and $Q$. This line intersects the curve at a third point $R$
```

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

>   - reflect the point R across the x-axis by changing the sign of the y-coordinate. The resulting point is P+Q.

Reviewer 3: ```suggestion
  - reflect the point $R$ across the $x$-axis by changing the sign of the $y$-coordinate. The resulting point is $P+Q$.
```

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

- if $P=Q$

>   - draw a tangent line throw P for an elliptic curve. The line will intersect the curve at the second point R.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

>   - reflect the point R across the x-axis the same way to get point 2P

Reviewer 3: ```suggestion
  - reflect the point $R$ across the $x$-axis the same way to get point $2P$
```

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

  - If the bit is 0, $y$ is positive; if 1, $y$ is negative. We'll consider the number positive by taking the smallest value between $y$ and $-y$: first compare $c_1$, then $c_0$.

The encoding of $x \in F_{p^2}$ as [u8; 96] bytes follows the rules from the section “Extension Fields Elements $F_{p^2}$”.

Encoding a point on $E'(F_{p^2})$ with a negative $y$ coordinate:

```rust
let x: [u8; 96] = encodeFp2(x);
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x20;
```

Encoding the point at infinity:

```rust
let x: [u8; 96] = [0; 96];
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x40;
```

#### ERROR_CODE

Validating the input for the host functions within the contract can consume significant gas.
For instance, verifying if a point belongs to the subgroup is gas-consuming. 
If an error is returned by the near host function, the entire execution is reverted. 
To mitigate this, when the input verification is complex, the host function 
will successfully complete its work but return an ERROR_CODE. 
This enables users to handle error cases independently. It's important to note that host functions
might terminate with an error if it's straightforward to avoid it (e.g., incorrect input size).

> The ERROR_CODE is encoded as a little-endian u64 and can hold the following values:

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8453fed5d464661da199f54339a34765f4986a06

This section aims to verify the correctness of summing up two valid elements on the curve:

- Utilize points on the curve with known addition results for comparison, such as tests from EIP-2537[^47],[^48].
- Generate random points on the curve and verify the commutative property: P + Q = Q + P.
- Validate that the sum of random points from $G_1$ remains in $G_1$.
- Generate random points on the curve and use another library to cross-check the results.

Edge cases:

- Points not from $G_1$.
- $\mathcal{O} + \mathcal{O} = \mathcal{O}$.
- $P + \mathcal{O} = \mathcal{O} + P = P$.
- $P + (-P) = (-P) + P = \mathcal{O}$.
- P + P (tangent to the curve).
- The sum of two points P and (-(P + P)) (tangent to the curve at point P).

<ins>Tests for inversion</ins>

This section aims to validate the correctness of point inversion:

- Generate random points on the curve and verify $P - P = -P + P = \mathcal{O}$.
- Generate random points on the curve and verify -(-P) = P.
- Generate random points from $G_1$ and ensure that -P also belong to $G_1$.
- Utilize an external implementation, generate random points on the curve, and compare results.

Edge cases:

- Point not from $G_1$

> - -$\mathcal{O}$

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8ea07dabdc798a37296877ede4a4a244559cfd70

- $e(P, Q + R) = e(P, Q) \cdot e(P, R)$
- $e(P + S, R) = e(P, R)\cdot e(S, R)$

Consequently:

$e([a]P, [b]Q) = e(P, Q)^{ab} = e([b]P, [a]Q)$

This function is necessary to verify BLS signatures/zkSNARKs.

This function takes as input the sequence of pairs $(p_i, q_i)$, where $p_i \in G_1 \subset E(F_{p})$ and $q_i \in G_2 \subset E'(F_{p^2})$ and validates:

$$
\prod e(p_i, q_i) = 1
$$

We don’t need to calculate the pairing function itself as the result would lie on a huge field, and in all known applications only this validation check is necessary.

***Input:*** A sequence of pairs $(p_i, q_i)$, where $p_i \in G_1 \subset E(F_{p})$ and $q_i \in G_2 \subset E'(F_{p^2})$. Each point is encoded in decompressed form. An expected input size of 288*k bytes is anticipated, interpreted as byte concatenation of k slices. Each slice comprises the concatenation of an uncompressed point from $G_1 \subset E(F_p)$ (occupying 96 bytes) and a point from $G_2 \subset E'(F_{p^2})$ (occupying 192 bytes). Additional details can be found in the Curve Points Encoding section.

***Output:***

The ERROR_CODE is returned.

- ERROR_CODE = 0: the input is correct

>   - <ins>Output:</ins> 1 byte with two possible values: 1 if the pairing result equals the multiplicative identity, and 0 otherwise.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/8453fed5d464661da199f54339a34765f4986a06

The implementation of  BLS12-381 curve operations from this NEP as a precompile will allows effective verify the BLS-signature and zkSNARKs. At the moment, BLS signature verification for BLS12-381 is impossible due to the limitation of the gas in 300 TGas for one transaction.

> Effective BLS-signature verification based on BLS12-381 elliptic curve will be useful for c*ross-chain interactions.* Some of the blockchains use the BLS signature in the protocols. If we want to implement the Clients for this blockchain on-chain in Near, we should be able to effectively verify the BLS signature. We can want to have a Client for a specific blockchain on Near for creating a bridge to verify the transaction from another blockchain and use it in Near. Examples of blockchains that use BLS signature based on BLS12-381: Eth2.0[[3](https://github.com/ethereum/consensus-specs/blob/master/specs/phase0/beacon-chain.md)], Filecoin[[6](https://spec.filecoin.io/)] and Tezos[[5](https://wiki.tezosagora.org/learn/futuredevelopments/layer2#zkchannels)]. Especially, it is necessary for Rainbow Bridge[[17](https://near.org/blog/eth-near-rainbow-bridge)] to make trustless transfers from Ethereum 2.0 to Near.

Reviewer 4: ```suggestion
Effective BLS-signature verification based on BLS12-381 elliptic curve will be useful for cross-chain interactions. Some of the blockchains use the BLS signature in the protocols. If we want to implement the clients for this blockchain on-chain in NEAR, we should be able to effectively verify the BLS signature. We can want to have a client for a specific blockchain on NEAR for creating a bridge to verify the transaction from another blockchain and use it in NEAR. Examples of blockchains that use BLS signature based on BLS12-381: Eth2.0[[3](https://github.com/ethereum/consensus-specs/blob/master/specs/phase0/beacon-chain.md)], Filecoin[[6](https://spec.filecoin.io/)] and Tezos[[5](https://wiki.tezosagora.org/learn/futuredevelopments/layer2#zkchannels)]. Especially, it is necessary for Rainbow Bridge[[17](https://near.org/blog/eth-near-rainbow-bridge)] to make trustless transfers from Ethereum 2.0 to NEAR.
```

Reviewer 4: The following should be replace in the document

- /s/Client/client
- /s/Near/NEAR (unless URL)

Author: fix: https://github.com/near/NEPs/pull/488/commits/664373f49804838b718ea8d6b6ea8a8d954246cd

Reviewer 5: @olga24912 I'd also replace client with "light client"

Author: fixed: https://github.com/near/NEPs/pull/488/commits/489ef31ad21868a061e74e1d230c03819a6e55fe

> zkSNARKs is useful for working with users' private information[[18](https://eips.ethereum.org/EIPS/eip-196),[19](https://media.consensys.net/introduction-to-zksnarks-with-examples-3283b554fc3b)]. Zeropool[[20](https://zeropool.network/)] is a project who implements zkSNARKs verifier on Near and is currently based on alt-bn128. Implementation of the precompiles for BLS12-381 can make the projects like that more secure. zkSNARKs is also used in Roll Ups[[21](https://www.youtube.com/watch?v=al4YpfDVmS4&ab_channel=EthereumCatHerders),[22](https://near.org/blog/layer-2),[23](https://www.ledger.com/academy/what-are-blockchain-rollups)] scaling solutions.

Reviewer 4: ```suggestion
zkSNARKs is useful for working with user's private information[[18](https://eips.ethereum.org/EIPS/eip-196),[19](https://media.consensys.net/introduction-to-zksnarks-with-examples-3283b554fc3b)]. Zeropool[[20](https://zeropool.network/)] is a project who implements zkSNARKs verifier on Near and is currently based on alt-bn128. Implementation of the precompiles for BLS12-381 can make the projects like that more secure. zkSNARKs is also used in Roll Ups[[21](https://www.youtube.com/watch?v=al4YpfDVmS4&ab_channel=EthereumCatHerders),[22](https://near.org/blog/layer-2),[23](https://www.ledger.com/academy/what-are-blockchain-rollups)] scaling solutions.
```

This proposal is based on a similar proposal for Ethereum: EIP-2537[[15](https://eips.ethereum.org/EIPS/eip-2537)].

In this NEP we propose to add the following functions as precompile:

> - ***bls12381_g1_sum*** — the function which adds the points from G1 on an elliptic curve. This function is useful for the aggregation of private keys in BLS Signature. Can be used for simple addition in G1. Separate from multiexp function due to gas cost. A similar function exists in Near for BN254 curve[[10](https://github.com/near/NEPs/issues/98)].
> - ***bls12381_g2_sum*** — the function which adds the points from G2 on an elliptic curve. This function is useful for the aggregation of signatures in BLS Signature. Can be used for simple addition in G2. Separate from multiexp function due to gas cost.

Reviewer 4: ```suggestion
- ***bls12381_g1_sum —*** the function which adds the points from G1 on an elliptic curve. This function is useful for the aggregation of private keys in BLS Signature. Can be used for simple addition in G1. Separate from multiexp function due to gas cost. A similar function exists in Near for BN254 curve[[10](https://github.com/near/NEPs/issues/98)].
- ***bls12381_g2_sum —*** the function which adds the points from G2 on an elliptic curve. This function is useful for the aggregation of signatures in BLS Signature. Can be used for simple addition in G2. Separate from multiexp function due to gas cost.
```

Keep document consistent, either all dashes highlighted or not.

- ***bls12381_g1_multiexp —*** for points $g_i \in G_1$ and scalars $s_i$ calculate $\sum g_i s_i$. Can be used for multiplication on the scalar. Can be useful for zkSNARKs verification. This operation can be performed in a more optimized way than just straightforward multiplication and addition by using Pippenger algorithm[[25](https://github.com/wborgeaud/python-pippenger/blob/master/pippenger.pdf)]. A similar function exists both in Near for BN254[[10](https://github.com/near/NEPs/issues/98)] and in EIP-2537[[15](https://eips.ethereum.org/EIPS/eip-2537)].

> - ***bls12381_g2_multiexp —***  for points $g_i \in G_2$ and scalars $s_i$ calculate $\sum g_i s_i$. Can be used for multiplication on the scalar.

Reviewer 4: ```suggestion
- ***bls12381_g2_multiexp —*** for points $g_i \in G_2$ and scalars $s_i$ calculate $\sum g_i s_i$. Can be used for multiplication on the scalar.
```

- ***bls12381_g1_map_to_curve —*** map base field element into the $G_1$ point. Doesn’t perform mapping of the byte string into field elements (can be done in different ways and quite fast). Transfer field element into a curve. It is necessary for signature schemes. Function from EIP-2537[[15](https://eips.ethereum.org/EIPS/eip-2537)].
- ***bls12381_g2_map_to_curve —*** map extension field element into the $G_2$ point. Doesn’t perform mapping of the byte string into extension field elements. Function from EIP-2537[[15](https://eips.ethereum.org/EIPS/eip-2537)].
- ***bls12381_g1_decompress —*** accepts points from $G_1$  in compressed form and returns in decompressed form. Some protocols provide points in compressed forms (for example, Light Client updates in Ethereum 2), and decompressing is a time-consuming operation. Other functions accept only decompressed points for simplicity and for gas consumption optimization.
- ***bls12381_g2_decompress —*** accepts points from $G_2$  in compressed form and returns in decompressed form.

> - ***bls12381_pairing —***  verifying that $\prod e(p_i, q_i) = 1$, where $e$ is a pairing operation and $p_i \in G_1 \land q_i \in G_2$. Necessary function for verification BLS-signatures or zkSNARKs.  A similar function exists both in Near for BN254[[10](https://github.com/near/NEPs/issues/98)] and in EIP-2537[[15](https://eips.ethereum.org/EIPS/eip-2537)].

Reviewer 4: ```suggestion
- ***bls12381_pairing —*** verifying that $\prod e(p_i, q_i) = 1$, where $e$ is a pairing operation and $p_i \in G_1 \land q_i \in G_2$. Necessary function for verification BLS-signatures or zkSNARKs.  A similar function exists both in Near for BN254[[10](https://github.com/near/NEPs/issues/98)] and in EIP-2537[[15](https://eips.ethereum.org/EIPS/eip-2537)].
```

In addition, there are implementations in other languages that are not so interesting to us in this context, but can be used as references:

1. C++, ETH2.0 Client, ***Chia library***[[37](https://github.com/Chia-Network/bls-signatures)]
2. Haskell, ***Adjoint Lib*** [[38](https://github.com/sdiehl/pairing)]
3. Go, ***Go-Ethereum*** [[39](https://github.com/ethereum/go-ethereum/tree/master/core/vm/testdata/precompiles)]
4. JavaScript, ***Noble JS*** [[40](https://github.com/paulmillr/noble-bls12-381)]
5. Go, ***Matter Labs Go EIP-1962 implementation*** [[41](https://github.com/kilic/eip2537)]
6. C++, ***Matter Labs Go EIP-1962 implementation*** [[42](https://github.com/matter-labs-archive/eip1962_cpp)]

> The draft implementation to nearcore you can find by this link[[54](near/nearcore#9317)]. This implementation is based on blst library[[30](https://github.com/supranational/blst)]. This library one of the fastest[[45](https://hackmd.io/@gnark/eccbench)] and audited[[55](https://research.nccgroup.com/wp-content/uploads/2021/01/NCC_Group_EthereumFoundation_ETHF002_Report_2021-01-20_v1.0.pdf)].

Reviewer 4: ```suggestion
The draft implementation to nearcore you can find by this link[[54](https://github.com/near/nearcore/pull/9317)]. This implementation is based on blst library[[30](https://github.com/supranational/blst)]. This library one of the fastest[[45](https://hackmd.io/@gnark/eccbench)] and audited[[55](https://research.nccgroup.com/wp-content/uploads/2021/01/NCC_Group_EthereumFoundation_ETHF002_Report_2021-01-20_v1.0.pdf)].
```

  - reflect the point R about the x-axis by changing the sign of the y-coordinate. The resulting point is P+Q.
- if $P=Q$
  - draw a tangent line throw P for an elliptic curve. The line will intersect the curve at the second point R.
  - reflect the point R about the x-axis the same way to get point 2P
- $P = -Q$
  - $P + Q = P + (-P) = 0$ — the point on infinity
- Q = 0
  - $P + Q = P + 0 = P$

With the addition operation, Elliptic Curve forms a **group**.

#### Subgroups

**Definition:**  Subgroup H is a subset of the group G with the following properties:

- $\forall h_1, h_2 \in H\colon h_1 + h_2 \in H$
- $0 \in H$
- $\forall h \in H \colon -h \in H$

Notation: $H \subseteq G$

**Definition:** group/subgroup **order** is the number of elements in group/subgroup.

Notation: |G|  or #G, where G is group

For some technical reason (for `pairing` operation which we will define later), we will work not with the hole $E(F_p)$, but only with the two subgroups $G_1$ and $G_2$ with the same **order** $r$. $G_1$ is a subset of $E(F_p)$, $G_2$ is a subgroup of another group, which we will define later. The $r$ should be prime and $G1 \ne G2$

> 

Reviewer 6: Typo: Replace "hole" by "whole"

Author: Thanks!

fixed: https://github.com/near/NEPs/pull/488/commits/4b225f0fd19c45de88e5e7cff272bd6452994b9f

- $$k_{(4,1)} = 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559571 \cdot I$$

- $$k_{(4,2)} = 18 + 4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559769 \cdot I$$

**Step 3:** Effective cofactor:

- $$h_{eff} = 209869847837335686905080341498658477663839067235703451875306851526599783796572738804459333109033834234622528588876978987822447936461846631641690358257586228683615991308971558879306463436166481$$

### Curve points encoding

#### Bool as output

The bool is encoded as a little-endian `u64` type in Rust. The `true` value is encoded as `1` and `false` is encoded as `0`. All other values of `u64` are not allowed and should be interpreted as incorrect. Encoding is the same as for BN254 implementation in nearcore[^50].

#### Sign

The sign of the point on the elliptic curve is encoded as `u8` type in Rust with two possible values: `0` and `1`. `0` for positive sign, and `1` for negative sign. All other value of `u8` is not allowed and should be interpreted as incorrect.

#### Scalar

> The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].

Reviewer 3: I believe that this particular encoding was chosen to be consistent with existing standards. Otherwise, it would be necessary to re-encode any actual inputs (reverse the order of bytes, set/clear some bits, etc.), and the current approach seems to be more developer-friendly.

> The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].

Reviewer 2: I guess it’s a matter of choice between being consistent with existing standards, and being consistent with our own codebase. Considering how I regard PHP, I’d lean towards being consistent with our own codebase rather than being consistent with external habits, but I can understand opinions vary on this matter.

> The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].

Reviewer 6: I agree with @abacabadabacaba that in this case it might be better to be consistent with existing standards, especially given that one of the goals of this NEP is to improve interoperability.

> The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].

Author: @Ekleog-NEAR Actually, I just double checked it and it really should be a little-endian. Standart which I whould like to follow says nothing about scalars (https://github.com/supranational/blst#serialization-format). In some libraries it is possible to decode both from le and be. And blst is using le for scalars, just checked it. 

I will fix this place. Thank you for notice.

> The scalar value is encoded as a big-endian `[u8;32]`. All possible bytes combination is allowed. Encoding is similar with BN254 implementation in nearcore[^50], but `big-endian` encoding is used instead of `little-endian` as in EIP-2537[^15].

Author: fixed: https://github.com/near/NEPs/pull/488/commits/7c5cb3b4c8ffea0d13f0a0ec6127923f288fc9dc

- $c_1 \in F_p$ `[u8; 48]`
- $c_0 \in F_p$ `[u8; 48]`

The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].

#### Uncompressed points on curve E(Fp)

The points on the curve represent by affine coordinates: `(x: Fp, y: Fp)`. The elements from $E(F_p)$ encoded in `[u8; 96]` as bytes’ concatenation of `x` and `y` point coordinates, where $x, y \in F_p$. The $x, y$ are encoded according to the rules described in the section “Fields elements $F_p$” .

$E(F_p)$  is encoded as `[u8; 96]`:

- $x \in F_p$ `[u8; 48]`
- $y \in F_p$ `[u8; 48]`

*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.

Encoding point on infinity:

```bash
let x: [u8; 96] = [0; 96];
x[0] = x[0] | 0x40;
```

The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].

#### Compressed points on curve E(Fp)

The points on the curve represent by affine coordinates: `(x: Fp, y: Fp)`. The elements from $E(F_p)$ in compressed form are encoded in `[u8; 48]` as *big-endian* encoded $x \in F_p$. The $y$  coordinate can be detected by the formula: $y = \pm \sqrt{x^3 + 4}$.

> *The first bit* should be set as 1. This bit indicates that point is encoded in compressed form.

Reviewer 3: Similar to the above, this makes it consistent with the existing standards. It is debatable whether it makes sense to use a nonstandard but "better" format (i.e. little endian, without extra bits, etc.), this NEP proposes to stick to the standard.

> *The first bit* should be set as 1. This bit indicates that point is encoded in compressed form.

Author: @Ekleog-NEAR 

I want to follow some existing encoding format and not develop my own. There are a few different candidates for that. For example, first I thought to encode it in the same way as in EIP-2537. 

But in EIP-2537 only uncompressed points are used. In this NEP we have functions that take the compressed points (decompress functions). As a result, we should have an encoding format for these points as well. 

In the end, I decided to follow ZCash encoding format: https://github.com/supranational/blst#serialization-format

This format is used in different BLS libraries(blst, milagro and so on) and it is also used for encoding public keys, and signatures in BLS-signatures in Ethereum. 

And I prefer to follow exactly this format even if it has some redundancy. I use this format to be compatible with different libraries and applications.  

Also, if we follow this format it will reduce the bugs' probability in implementation. Because in that case, we need not to re-encode data. 

Most probably this bit will be set anyway because the data were serialized by using some BLS library and if the bit is not set, more likely that the data is incorrect.

*The second bit* is used to mark a point on infinity, if the second bit is set to 1 — it is an infinity point.

To represent the sign of the $y$ *the third bit* of x encoding is used. If the first bit = 0 the $y$ is positive, if the first bit = 1 the $y$ is negative. We will consider the number positive the smallest between $y$  and $-y = p - y$.

The $x \in F_p$ encoded as `[u8; 48]` bytes according to the rules from section  “Extension fields elements $F_{p}$” .

The point on $E(F_p)$ with negative $y$ coordinate encoded as `[u8; 48]` bytes:

```rust
let x: [u8; 48] = encodeFp(x)
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x20;
```

For encoding point of infinity:

```rust
let x: [u8; 48] = [0; 48];
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x40;

> ```

Reviewer 3: Elliptic curve may have a single point at infinity (not positive or negative, just infinity). This code would produce encoding where all bits other than the first two are zero (because of `[0; 48]`), and I think that the code should reject any other encoding.

> ```

Reviewer 6: In this [implementation](https://github.com/zkcrypto/pairing/blob/34aa52b0f7bef705917252ea63e5a13fa01af551/src/bls12_381/README.md) which is used in several places, they also state that, if the bit for the point is at infinity is set, then the remaining bits of the group element's encoding should be set to zero, which matches the suggestion above.

> ```

Author: @Ekleog-NEAR Yes, all other bits should be set to 0.  I will write it more explicitly and also will add a test case with an incorrect infinity encoding.

> ```

Author: fix: https://github.com/near/NEPs/pull/488/commits/90cb55f4a0dbd07f7ff794939197cddd3e36c5b6

The $x \in F_{p^2}$ encoded as `[u8; 96]` bytes according to the rules from section  “Extension fields elements $F_{p^2}$” .

The point on $E'(F_{p^2})$ with negative $y$ coordinate encoded as `[u8; 96]` bytes:

```rust
let x: [u8; 96] = encodeFp2(x);
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x20;
```

Encoding point of infinity:

```rust
let x: [u8; 96] = [0; 96];
x[0] = x[0] | 0x80;
x[0] = x[0] | 0x40;
```

The rule of encoding is consistent with zkcrypto[^53] and with implementation in milagro lib[^29].

### Host functions

#### bls12381_g1_sum

***Description:***

The function computes the sum of the signed elements of the BLS12-381 curve. The input is an arbitrary number of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point on elliptic curve and $s_i \in \textbraceleft 0, 1 \textbraceright$ is the point sign. The output is one point from $E(F_p)$ equal to $\sum (-1)^{s_i}p_i$.

The $E(F_p)$ curve,  points on the curve, multiplication on -1,  and the addition operation are defined in the BLS12-381 Curve Specification section.

Note: we take as input any points on the curve, not only from   $G_1$

> ***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point and $s_i \in \textbraceleft 0, 1 \textbraceright$ is sign, each point encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and sign encoded in one byte with only two allowed values: 0, 1. Expected `97*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the uncompressed point from $E(F_p)$ and a bool value for point sign. More details are in the Curve Points Encoding section.

Author: Yes, I have done it in the same way as in bn254 implementation. Also in our multiexp functions, scalars go after points. I just try to make it similar. If someone wants to switch from bn254 to bls12-381 it can be convenient to have the same ABI. 

I don't have a strong opinion about what is the right way. From the implementation point of view, it is no different.

> ***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point and $s_i \in \textbraceleft 0, 1 \textbraceright$ is sign, each point encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and sign encoded in one byte with only two allowed values: 0, 1. Expected `97*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the uncompressed point from $E(F_p)$ and a bool value for point sign. More details are in the Curve Points Encoding section.

Reviewer 2: Uhhh… in my copy of logic.rs, I see ``value` - sequence of (sign:bool, g1:G1), where [...]` as a comment on `alt_bn128_g1_sum`, so the sign is before. Am I missing something?

> ***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point and $s_i \in \textbraceleft 0, 1 \textbraceright$ is sign, each point encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and sign encoded in one byte with only two allowed values: 0, 1. Expected `97*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the uncompressed point from $E(F_p)$ and a bool value for point sign. More details are in the Curve Points Encoding section.

Author: Hmm... I see, it is not consistent with https://github.com/near/NEPs/issues/98.

Ok, I will change it. Thanks for noticing it.

> ***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is point and $s_i \in \textbraceleft 0, 1 \textbraceright$ is sign, each point encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and sign encoded in one byte with only two allowed values: 0, 1. Expected `97*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the uncompressed point from $E(F_p)$ and a bool value for point sign. More details are in the Curve Points Encoding section.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/d9b039a05f6552e8005ebd4102c032def6cf4226

***Output:*** the output is 96 bytes — the one point $\in E(F_p)$ in decompressed form. For empty input it returns point on infinity (see  Curve Points Encoding section).

***Gas Estimation:***

The algorithm has a linear complexity of the number of elements. The gas can be calculated by the next formula

```rust

> let k = (input_bytes + item_size - 1)/item_size

Author: fixed: https://github.com/near/NEPs/pull/488/commits/b737d0b80918b055c07ea9d12094e893e5ad5dcc

Note:

- We take as input any points on the curve, not only from   $G_1$.
- The scalar is an arbitrary unsigned integer and can be bigger than the group order.

***Input:*** the sequence of pairs $(p_i, s_i)$, where $p_i \in E(F_p)$ is a point on the curve and $s_i \in \mathbb{N}_0$ is a scalar. Each point is encoded in decompress form as $(x\colon F_p, y\colon F_p)$ and a scalar has a type `u256` and BigEndian encoded in 32 bytes. Expected `128*k` bytes as an input that is interpreted as byte concatenation of `k` slices, each slice is the concatenation of uncompressed point from $E(F_p)$ — `96` bytes and scalar — `32` bytes. More details are in the Curve Points Encoding section.

***Output:*** the output is 96 bytes — the one point $\in E(F_p)$ in decompressed form. For empty input it returns point on infinity (see  Curve Points Encoding section).

***Gas Estimation:***

This function should be calculated by Pippenger’s algorithm[^25]. The Complexity of this algorithm is $O(\frac{k}{\log(k)})$. For gas calculation we will use the formula $\frac{k}{\max(\log_2(k), 1)}$ the same way as in the host function for `BN254`[^10].

```rust
let k = (input_bytes+item_size-1)/item_size;

> let gas_consumed = A + B*k + C * if k > 1 {k / (k as f32).log2().floor()} else {k};

Author: @Ekleog-NEAR From the algorithmic point of view, it is a correct formula. I took this from bn254 NEP. However, after I started implementing this, and figured out, that for bn254 this formula is not implemented, instead just a linear formula is used. 

The idea of that formula: the algorithm gives some speed boost and for using this boost we need this formula. Otherwise, we will charge too much gas on the big number of points. 

In EIP-2537 they have just a list of constants of how much gas they should charge for different N. 

Now, I have started to think, that the best solution:
- using the linear formula for gas calculation
- don't specify which algorithm should be used  in the implementation

The Pippenger algorithm is implemented only in the master branch of blst library, not in the release version. And I started to doubt that implementation is safe. As a result, I think to make the first implementation as simple as possible and don't use a clever algorithm. In the future, if we will need it, we can update it to the faster version. We can update it without changing this NEP. And I expect that for this function, slow implementation will be acceptable.

> let gas_consumed = A + B*k + C * if k > 1 {k / (k as f32).log2().floor()} else {k};

Author: fixed: https://github.com/near/NEPs/pull/488/commits/80adcc7167832afd653386381d621b77a01d1f4f

```

A, B and C are constants calculated empirically.

For gas estimation, we can use the benchmark vectors for addition and multiplication for EIP-2537[^46].

***Test cases:***

The same test cases as for the bls12381_g1_sum section.

Addition test cases:

- `group_order * P = 0`
- `(scalar + groupt_order) * P = scalar * P`
- `P + P + P .. + P = N*P`
- `0 * P = 0`
- `1 * P = P`
- Scalar is a MAX_INT

***Tests References:***

The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].

***Error cases:***

- The input length is not divided by 128
- Too much memory is used
- Field elements encoded incorrectly (see Curve points encoded section)
- Any points not on the curve

***Annotation:***

```rust
/// Computes multiexp on BLS12-381 curve using Pippenger's algorithm 
///\sum_i si*pi should be equal result.
///
/// # Arguments
///
/// * `value` - sequence of (pi, si), where
///    pi is point (x:Fp, y:Fp) on BLS12-381, and si is u256.
///    BLS12-381 is Y^2 = X^3 + 4 curve over Fp.
///
///   `value` is encoded as packed, big-endian

> ///   `[(([u8; 48], [u8; 48]), u256)]` slice.

Reviewer 3: `value` is encoded as a concatenation of `(x, y, s)` tuples, where `x: [u8; 48]`, `y: [u8; 48]` and `s: [u8; 32]` encode a point and a scalar.

> ///   `[(([u8; 48], [u8; 48]), u256)]` slice.

Author: @Ekleog-NEAR I'll make the explanation clearer

> ///   `[(([u8; 48], [u8; 48]), u256)]` slice.

Author: just removed this part, because comments in the annotation don't add any new information: https://github.com/near/NEPs/pull/488/commits/cc448df84cb9d3a3d4dd73b14097b9ef011f2d6d

***Input:*** the function takes as input `48` bytes — the element from $F_p$ (one unsigned integer $< p$). More details are in the Curve Points Encoding section.

***Output:*** the output is `96 bytes` — one point $\in G_1 \subset E(F_p)$ in decompressed format. More details are in the Curve Points Encoding section.

***Gas Estimation:***

The gas consumption is a constant calculated empirically.

***Test cases:***

- Correct $F_p$ element
- $a = 0$
- $a \ge p$
- Edge cases for inner algorithms for mapping[^49]

***Tests References:*** The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].

***Error cases:***

- Incorrect input length
- $a \ge p$

***Annotation:***

```rust
/// Map elements from Fp to G1 subset of BLS12-381 curve
///
/// # Arguments
///
/// * `value` - 48 bytes, the element from Fp -- unsigned integer < p 
///
/// # Errors
///
/// If `value_len + value_ptr` points outside the memory or the registers
/// use more memory than the limit, the function returns
/// `MemoryAccessViolation`.
///

> /// If value_len != 48 or the value >= p the function return `BLS12381InvalidInput`.

Reviewer 2: Why have a parameter, if it’s to hardcode it to a single valid value? Do you foresee extensions where this value would change?

Reviewer 3: Because if the value comes from a register, then `value_len` would be `u64::MAX`, and the code would use the length of the value in the register instead.

Reviewer 2: Wait, is values coming from registers part of the specification proposed? As far as I can tell this function needs a `value_ptr`, which means it must be somewhere in the memory of the wasm program.

Reviewer 3: Generally NEAR host functions obtain input from memory by using `get_memory_or_register!`, which can also get it from a register.

Reviewer 2: This is "generally", if we do want it it needs to be explicitly listed in the NEP. And I agree that we probably do want it considering our other host functions, but the current NEP, as written, must not have an implementation that uses `get_memory_or_register`, as it never mentions being able to take input from registers.

Author: @Ekleog-NEAR Hm... The difficult conversations are started. Yes, in implementation I am going to use `get_memory_or_register`. It would be good if you give me an example of how it is explained in another NEP if you have something in your mind. If you don't have a quick answer -- no problem, I will find an example by myself.

Reviewer 2: Unfortunately I think we only recently started to actively use formalized NEP, so I don’t know of a good example. I’d say just a short sentence in the definition of each function that takes address+len, that says that if len is u64::max then registers are used to fetch the data, and the length of the register is used, should I think be enough to be explicit that ptr+len is not only used to refer to the memory address space :)

***Input:*** the function takes as input `96 bytes` — the element from $F_{p^2}$ (two unsigned integers $< p$). More details are in the Curve Points Encoding section.

***Output:*** the output is `192 bytes` — one point $\in G_2 \subset E'(F_{p^2})$ in decompressed format. More details are in the Curve Points Encoding section.

***Gas Estimation:***

The gas consumption is a constant calculated empirically.

***Test cases:***

- Correct $F_{p^2}$ element
- $a = 0$
- One of the `a` value $\ge p$
- Edge cases for inner algorithms for mapping[^49]

***Tests References:*** The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].

***Error cases:***

- Incorrect input length
- Value is not a valid extension field $F_{p^2}$ element

***Annotation:***

```rust
/// Map elements from Fp2 to G2 subset of E'(Fp2) BLS12-381 curve
///
/// # Arguments
///
/// * `value` - 96 bytes, the element from Fp2 -- two unsigned integer < p 
///
/// # Errors
///
/// If `value_len + value_ptr` points outside the memory or the registers
/// use more memory than the limit, the function returns
/// `MemoryAccessViolation`.
///

> /// If value_len != 96 or the value is not correct Fp2 element the function return `BLS12381InvalidInput`.

Reviewer 2: (bis)

***Tests References:*** The EIP-2537 contains the same function, so we can reuse test vectors from Ethereum[^47],[^48].

***Error cases:***

- The input length is not divided by 288
- Too much memory is used
- Field elements encoded incorrectly (see Curve points encoded section)
- Any points not on the curve
- Any points not in $G_1/G_2$

***Annotation:***

```rust
/// Computes pairing check on BLS12-381 curve.
/// \prod_i e(g_{1 i}, g_{2 i}) should be equal one, e(g1, g2) is pairing
///
/// # Arguments
///
/// * `value` - sequence of (g1:G1, g2:G2), where
///   G2 is subgroup point (x:Fp2, y:Fp2) on BLS12-381 twist,
///   BLS12-381 twist is Y^2 = X^3 + 4(i + 1) curve over Fp2
///   Fp2 is complex field element (re: Fp, im: Fp)
///   G1 is point (x:Fp, y:Fp) on BLS12-381,
///   BLS12-381 is Y^2 = X^3 + 4 curve over Fp
///
///   `value` is encoded a as packed, big-endian
///   `[(([u8; 48], [u8; 48]), ([u8; 96], [u8; 96]))]` slice.
///
/// # Errors
///
/// If `value_len + value_ptr` points outside the memory than
/// the function returns `MemoryAccessViolation`.
///
/// If point coordinates are not on curve, point is not in the subgroup
// or data are wrong serialized, for example,

> /// `value.len()%288!=0`, the function returns `BLS12381InvalidInput`.

Reviewer 2: Seems like this is by analogy with current bn128 host functions. It makes me sad (we should have just multiplied by 288 at the beginning of the host function), but I guess consistency is more important than me being happy in this case, so it probably shouldn’t change :’(

Reviewer 3: This doesn't work if the value comes from a register. Maybe we should have never supported using registers to pass input to host functions, but now AFAIK all existing host functions support this.

```

## Reference Implementation

First of all, for integration with nearcore, we are interested in libraries in the Rust language. The existing BLS12-381 implementations on Rust:

1. ***Milagro Library*** [^29].
2. ***BLST***  [^30][^31].
3. ***Matter labs EIP-1962 implementation*** [^32]
4. ***zCash origin implementation*** [^33]
5. ***MCL Library*** [^34]
6. ***FileCoin implementation*** [^35]
7. ***zkCrypto*** [^36]

To compile the list, we used the links from EIP-2537[^43], pairing-curves specification[^44], and an article with benchmarks[^45]. This list may be incomplete but should cover the core BLS12-381 implementations. In any case, to implement host functions from that NEP we will need to modify any of that libraries.

In addition, there are implementations in other languages that are not so interesting to us in this context, but can be used as references:

1. C++, ETH2.0 Client, ***Chia library***[^37]
2. Haskell, ***Adjoint Lib***[^38]
3. Go, ***Go-Ethereum***[^39]
4. JavaScript, ***Noble JS***[^40]
5. Go, ***Matter Labs Go EIP-1962 implementation***[^41]
6. C++, ***Matter Labs Go EIP-1962 implementation***[^42]

> The draft implementation to nearcore you can find by this link[^54]. This implementation is based on *blst* library[^30]. This library one of the fastest[^45] and the most audited[^55].

Reviewer 2: I think this doesn’t need to be in the NEP. I personally don’t think an audited assembly library is necessarily safe (as proved by openssl still regularly having vulnerabilities), but I haven’t checked alternatives and this discussion is for another time anyway, when thinking about landing the implementation :)

Author: @Ekleog-NEAR 

What exactly should not be in the NEP? I shouldn't give my opinion about which library is better to use? 

I thought, that the opposite, NEP should include the list of libraries and draft implementation. 

Yes, that is true, if the library is audited it doesn't mean, that it is correct. And I haven't done any analytics to choose the best library.

However, according to some evidence(such as it is quite well known, used in Ethereum, and has some audits), this library is not a bad choice.  And, another important thing: it is possible to implement these host-functions using this library. 

The message of this paragraph: the host functions from this NEP easy to implement by using existing not bad* libraries.    

We are not obliged to use this particular library and can choose something else if we have a reason for it

*not bad = works acceptably fast, the library is not dead, used by other big projects

Reviewer 2: Right, I was just thinking about the "This library one of the fastest[^45] and the most audited[^55]." sentence; but it’s not a big deal anyway :) (I’m also not objecting to using this library for the production implementation)

Author: fixed: https://github.com/near/NEPs/pull/488/commits/1526529bc7559a75030f3b62a0d3bc73419f7b7c

The next alternative is to execute BLS12-381 operations off-chain. In that case, the applications which are using the BLS curve will not be trustless anymore.

## Future possibilities

In the future, it is possible to support work with other curves, not only BLS12-381. In Ethereum, before EIP-2537[^15], EIP-1962 was proposed[^27]. In EIP-1962 was proposed to implement pairing-friendly elliptic curves in a generic format and support not only BLS curves but many others. However, this proposal wasn't accepted due to its large scope and complexity. I don't think it makes sense to implement every possible curve, but it could be a possible extension.

## Consequences

### Positive

- Projects, which are using BN254 will be able to switch on BLS12-381 curve, and it will improve their security.
- The trustless cross-chain interactions with blockchains using BLS12-381 in protocols (such as Ethereum 2.0) will become possible.

### Neutral

### Negative

> - The appearance of dependence on the library which supports BLS12-381 curves operations.

Reviewer 2: Also add the fact that we must maintain a bls12-381 implementation forever, even when it becomes broken.

(This is one thing often reproached to OpenPGP, especially its support of 3DES or old mac schemes)

Reviewer 3: There is an important difference between how cryptographic systems were designed in 1990s vs today.
In 1990s, it was normal to pack many different algorithms into a system so that if some of them turn out to be broken, others hopefully remain secure. Protocols like TLS and OpenPGP hail from that era.
Nowadays, the best practice is to only support one or a very small number of algorithms, to make sure that none of them is broken. It turned out that even a single broken algorithm makes the entire system much weaker, because an attacker might be able to make the system fall back to using an insecure algorithm, or the user might be using it accidentally. Modern systems like [libsodium](https://en.wikipedia.org/wiki/Libsodium), [AGE](https://github.com/FiloSottile/age), [Double Ratchet](https://en.wikipedia.org/wiki/Double_Ratchet_Algorithm) and others follow this principle.
NEAR has to implement multiple cryptographic primitives in order to interoperate with other systems (such as Ethereum), but we evaluate each one separately, so hopefully we won't end up in OpenPGP situation.

Reviewer 2: I totally agree that we hopefully won’t end up in a bad situation (regardless of whether the OpenPGP situation is bad), and this is why I’m not saying this negative point should be blocking. But it should be explicitly listed, because we ought to be lucid about what the drawbacks of our NEPs are.

Author: @Ekleog-NEAR Sure, I will add this disadvantage

Author: fixed: https://github.com/near/NEPs/pull/488/commits/6e9c7b0728791b93fea92fb18ea13f54c6acc496

### Map to curve specification

This section explains the functionality of the `bls12381_map_fp_to_g1` and `bls12381_map_fp2_to_g2` functions. 
They operate based on the specification RFC9380 "Hashing to Elliptic Curves"[^62].

> This functions maps the field element($F_p$ or $F_{p^2}$) to the subgroup of elliptic curve($G_1 \subset E(F_p)$ or $G_2 \subset E'(F_{p^2})$).

Reviewer 6: Suggestion: These functions map a field element in $F_p$ or $F_{p^2}$ to the corresponding subgroup $G_1 \subset E(F_p)$ or $G_2 \subset E'(F_{p^2})$.

Author: fixed: https://github.com/near/NEPs/pull/488/commits/60a4a8099e1c89c25343bcd656fba575ad51a422
